---
title: "Microbial Community 2019 Analysis"
author: "Ricardo Silva"
date: "02/06/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval=FALSE, 
                      message=FALSE, 
                      warning=FALSE)
```

```{r CLEAR EVERYTHING, eval=FALSE, include=FALSE}
# unload all non-base packages
lapply(names(sessionInfo()$otherPkgs), function(pkgs)
  detach(
    paste0('package:', pkgs),
    character.only = T,
    unload = T,
    force = T
))

rm(list=ls())
```

# Packages and initial steps

```{r load libraries, message=FALSE, warning=FALSE, include=FALSE}
# load the packages
# main packs to start with
pcks <- c('phyloseq', 'microbiome','microbiomeutilities' ,'vegan', 'factoextra',"FactoMineR", "patchwork",'corrr','GGally','tidymodels',"ggforce", 'tidyverse', "circlize")

if(sum(as.numeric(!pcks %in% installed.packages())) != 0){
  installation <- pcks[!pcks %in% installed.packages()]
  for(i in 1:length(installation)) {
    install.packages(installation, dependencies = T)
    break()}
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
} else {
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
}

rm(pcks)
```

```{r packages and functions}
# set working directory and seed
setwd("C:/Users/rrocha/Documents/R/phd/2019/data")
folder_path <- "C:/Users/rrocha/Documents/R/phd/2019/output/"

source('~/R/phd/2019/script/my_functions.R')
source('~/R/phd/2019/script/theme_publication.R')
theme_set(theme_Publication_3())

# set colors
color_distance <- c("D0" = "deepskyblue2", 'D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# color_distance1 <- c('D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
color_depth <- viridis::viridis(n=3)
cage_control <- friendly_cols

# check the working directory
current_directory <- getwd()
paste("Working in directory:", current_directory)

# dir.create(paste0(folder_path, "plots/community_18s"))
```

# Aims

Asses protist community and function changes and check if 16S and functional profiling can predict nutrient concentrations

Here, we characterized how fish farms affect the spatial distribution of marine microorganisms in a semi-enclosed coastal environment, tested how similar the spatial niche of closely related taxa is, and what are the environmental parameters modulating their spatial abundance patterns.

We collected surface seawater samples (at a depth of 50 cm) in fish-net cages (29° 32′20” N, 121° 45′10” E, with suffix “F”) and an adjacent control site (29°36′4” N, 121° 46′10” E, 7 km apart from the fish-net cages, with suffix “C”)

# Load Data 

```{r}
# load object in RData format
# 18S
load("pseq.euk.list.RData")
pseq.euk.list
load("pseq.euk.list.filt.RData") # filtered data
pseq.euk.list.filt
# load("pseq.euk.list.clr.RData") # clr transformed data
# pseq.euk.list.clr

# meta
meta <- read_csv(paste0(getwd(),"/meta_impute.csv")) 


```


## Map - Sampling Sites

```{r}
# map_dpeth_profile.Rmd
```

## Community Analysis 

We compared community composition in seawater samples from each depth and distance from the cage. Statistical analysis of ASVs data was conducted principally in R [38] using the Vegan [39], Phyloseq, ggplot2, microbiome and Codaseq packages. 
Gloor et al. argue that microbiome datasets generated by high-throughput sequencing are compositional in nature because the number of DNA sequence reads is limited by the capacity of the sequencing machinery. Thus we analyzed this dataset compositionally by first filtering with CodaSeq (min.reads = 2000, min.occurrence = 0.001, min.prop = 0.001), then conducting a center log-ratio transformation (clr), instead of using standard counts and rarefying.

```{r number of taxa}
pseq.euk.list %>% map_df(., ~ tax_table(.) %>% 
                       as.data.frame() %>%
                       summarise(across(where(is.character), ~n_distinct(.x)))
                     ) %>%
  add_column(Community = names(pseq.euk.list))

pseq.euk.list.filt %>% map_df(., ~ tax_table(.) %>% 
                       as.data.frame() %>%
                       summarise(across(where(is.character), ~n_distinct(.x)))
                     ) %>%
  add_column(Community = names(pseq.euk.list))

```


### Summary tables

```{r summary}
# create file directory
# dir.create(paste0(folder_path,"data/summary_tables"))

summary_tables <- function(pseq.list, group, taxo){
 
  df <- pseq.list %>% 
    map(., ~subset_samples(., site == group)) %>%
    map2(.,names(.),~summarize_taxa(., taxo) %>% 
    arrange(desc(meanRA), .data[[taxo]]) %>%
      ungroup() %>%
  slice_max(meanRA, n = 5) %>%
      add_row(!!(taxo) := paste(.y), .before = 1) %>%
    mutate(Rank = taxo) %>%
     rename(taxa_id = taxo)) %>%
    bind_rows() 
  
  return(df)

}

group = 'Control'
control_phylum <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Phylum")
control_class <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Class")
control_order <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Order")
control_family <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Family")
control_genus <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Genus")
control_spp <- summary_tables(pseq.euk.list.filt, group = 'Control', taxo = "Species")

bind_rows(control_phylum, control_class, control_order, control_family, control_genus, control_spp) %>% 
  write_csv(paste0(folder_path, "data/summary_tables/summary_control.csv"))

group = 'lease_1'
lease1_phylum <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Phylum")
lease1_class <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Class")
lease1_order <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Order")
lease1_family <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Family")
lease1_genus <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Genus")
lease1_spp <- summary_tables(pseq.euk.list.filt, group = 'lease_1', taxo = "Species")

bind_rows(lease1_phylum, lease1_class, lease1_order, lease1_family, lease1_genus, lease1_spp) %>% 
  write_csv(paste0(folder_path, "data/summary_tables/summary_lease1.csv"))

group = 'lease_2'
lease2_phylum <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Phylum")
lease2_class <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Class")
lease2_order <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Order")
lease2_family <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Family")
lease2_genus <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Genus")
lease2_spp <- summary_tables(pseq.euk.list.filt, group = 'lease_2', taxo = "Species")

bind_rows(lease2_phylum, lease2_class, lease2_order, lease2_family, lease2_genus, lease2_spp) %>% 
  write_csv(paste0(folder_path, "data/summary_tables/summary_lease2.csv"))


# Genus
gen.sel <- pseq.euk.list.filt %>% 
  summarize_taxa("Genus", GroupBy = "site") %>% 
  group_by(site) %>% 
  mutate(freq_by_group = str_remove(freq_by_group, "%"),
         freq_by_group = as.numeric(freq_by_group)) %>%
  arrange(desc(freq_by_group), .by_group = T)  %>%
#   filter(Genus!="unknown") %>% 
    slice_max(freq_by_group, n = 10) %>% 
  distinct(Genus) %>% 
  pull
```

```{r}
# # look in more details
# pseq.list.filt[[2]] %>%
#   get_tibble('tax_table') %>%
#   filter(if_any(everything(), ~str_detect(.,"Arcobacteraceae"))) %>% distinct(Genus)
# 
# pseq.list.filt[[2]] %>% 
#   summarize_by_subtaxa(taxa = "Campylobacteria", Rank = "Genus", GroupBy = "site")  %>%
#     select(-Phylum) %>%
#   group_by(site) %>% 
#   mutate(freq.within.group = str_remove(freq.within.group, "%"),
#          freq.within.group = as.numeric(freq.within.group)) %>%
#    arrange(desc(freq.within.group), .by_group = T) %>% view
```

### Circus plot

```{r prep for circus plot 1}
library("circlize")
# set taxa level and phyloseq object according to Kingdom
Rank = "Family" # to group
Rank2 = "Order" # to add name (above the Rank)
GroupBy = "Habitat_Layer"
physeq = pseq.filt.list[[3]] # change the number (1-3)

ps.melt <- physeq %>%
  #tax_glom(taxrank = Rank) %>%                     # agglomerate at 'Rank' level
  psmelt.dplyr() %>%                                         # Melt to long format
  arrange(Rank)                                     # arrange by 'Rank'
#  columns <- c(Rank,GroupBy)

circus.df <- ps.melt %>%  
 # filter(Class == "Stramenopiles") %>%
  # if choose phylum as Rank, skip this line
  mutate("{Rank}" := paste(.data[[Rank2]], .data[[Rank]], sep = "_")) %>%
  # group by Rank and Group
  # {{Rank}} for function - to wrap the function arguments 
  # .data[[Rank]] for string inputs
  group_by(.data[[GroupBy]], .data[[Rank]]) %>%
  # summarise and create a column with the relative abundance
  dplyr::summarise(Abundance = sum(Abundance)) %>%
  pivot_wider(.data[[Rank]], names_from = .data[[GroupBy]], values_from = Abundance) %>%
  # reduce function combines (reduces) all of the elements into a single object `x` = sum (e.g. reduce(c(1, 2, 3), `+`))
  mutate(total = purrr::reduce(select(., -.data[[Rank]]), `+`)) %>% 
  arrange(desc(total)) 

# get to top 9 in total abundance
#top <- circus.df  %>% slice_max(total, n= 9) 
top <- circus.df  %>% slice_max(total, n= 15) 

# make another data frame with the others taxa
# checj funtion fct_lump -> mutate(species = fct_lump(species, n = 3))
no_top <- circus.df %>% anti_join(top) %>% # to make a complete dataframe use bind_rows()
  summarise(.,
        across(where(is.numeric), sum),
        across(where(is.character), ~"others"))

# add to the first dataframe
circus.df.plot <- bind_rows(top,no_top) %>%
# map function to map each column to the function you mention.
# map_at(c("a", "b", "c"), ~./sum(.)) %>% 
  map_if(is.numeric, ~./sum(.)) %>%
  as.data.frame()  %>%
  select(-total) %>%
  column_to_rownames({{Rank}})
```


```{r prep for circus plot - archaea}
# for Archaea only
circus.df.plot <- circus.df %>% 
  map_if(is.numeric, ~./sum(.)) %>%
  as.data.frame()  %>%
  select(-total) %>%
  column_to_rownames({{Rank}})
```


```{r circus plot table}
# save a table
kingdom <- ps.melt %>% select(Kingdom) %>% pull %>% unique
to_save <- circus.df.plot %>%
  rownames_to_column({{Rank1}}) %>%
          mutate(across(is.numeric, ~. *100),
          across(is.numeric, ~ round(.,2)))
write.csv(to_save, 
          paste0('C:/Users/rrocha/Documents/R/phd/2018/output/plots/all/composition/circus/circus_',{{kingdom}},"_",{{Rank2}},"_",{{Rank}},"_table",".csv"))
```


```{r circus plot}

# set colors
#library(RColorBrewer)
colourCount <- circus.df.plot %>% row.names() %>% n_distinct()
# colourCount <-length(unique(ps.glom[,length(ps.glom)]))
getPalette <- colorRampPalette(brewer.pal(colourCount, "Paired"))
grid.col <- circus.df.plot %>% rownames_to_column('taxa') %>%
  add_column(col = getPalette(colourCount)) %>%
  pull(col, taxa)
grid.col <- c(grid.col, cols_hab_layer)

# prep
mat <- as.matrix(circus.df.plot)

# save as svg 
svg(filename=, 
          paste0('C:/Users/rrocha/Documents/R/phd/2018/output/plots/all/composition/circus/circus_',{{kingdom}},"_",{{Rank}},".svg"), 
    width=11, # inches ( = x*144 pixels)
    height=6)

# SVG sizes are in inches, not pixels
#res <- 144
#svglite::svglite('C:/Users/rrocha/Documents/R/phd/2018/output/plots/all/composition/circus/test.svg', width = 1584/res, height = 864/res)

# set plotting area
par(xpd=NA) # or par(xpd=TRUE) to enable things to be drawn outside the plot region.

# Basic circus graphic parameters
circos.par(start.degree = 115, gap.after = c(rep(2, nrow(mat)-1), 10, rep(2, ncol(mat)-1), 10))

# script
chordDiagram(mat, grid.col = grid.col, directional = 1, direction.type = c("diffHeight", "arrows"),
             link.arr.type = "big.arrow", annotationTrack = c("grid"), scale = FALSE)

# axis and lables 
circos.track(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  xplot = get.cell.meta.data("xplot")
  circos.points(x = x, y = y, pch = "·")
  #plot labels
  circos.text(mean(xlim), y= 3 ,sector.name, facing = "reverse.clockwise", niceFacing = TRUE, adj = aa, cex = 1)
  #plot axis
  circos.axis(labels.cex=0.6, direction = "outside")
 
}, bg.border = NA)  # here set bg.border to NA is important

#par(mfrow=c(1,1)) # reset plotting area

dev.off()

# clean plotting area
circos.clear()

# testar
#https://stackoverflow.com/questions/26292484/programming-in-r-bubble-chart-visualization
#ggplot(test1, aes(x = origin, y = destination, size = frequencies)) + geom_point()
```

```{r circus plot 2}
# to analyze in more details use the strings bellow
# data prep 
Rank = "Order" # rank to be filtered
Rank1 = "Family" # to name (below the Rank)
GroupBy = "Habitat_Layer"
physeq = pseq.filt.list[[3]]
# let's say we want to look at the following taxa
to_filter = c("Ochrophyta")
to_filter = c("Proteobacteria|Bacteroidota|Actinobacteriota") # this way to filter 
taxa.to.filter = to_filter

ps.melt1 <- physeq %>%
  #tax_glom(taxrank = Rank1) %>%                     # agglomerate at 'Rank' level
  psmelt.dplyr() %>%                                         # Melt to long format
  arrange({{Rank}})                                     # arrange by 'Rank'

ps.melt <- ps.melt1 %>% filter(str_detect(.data[[Rank]], {{taxa.to.filter}}))
ps.melt <- ps.melt %>% unite({{Rank1}}, Rank, Rank1, sep = "_", remove = FALSE)

circus.df <- ps.melt %>%  
  # group by Rank and Group
  group_by(.data[[GroupBy]], .data[[Rank1]]) %>% 
  # summarise and create a column with the relative abundance
  dplyr::summarise(Abundance = sum(Abundance)) %>%
  pivot_wider(.data[[Rank1]], names_from = .data[[GroupBy]], values_from = Abundance) %>%
  # reduce function combines (reduces) all of the elements into a single object `x` = sum (e.g. reduce(c(1, 2, 3), `+`))
  mutate(total = reduce(select(., -.data[[Rank1]]), `+`)) %>% 
  arrange(desc(total)) 

# choose the top taxa in total abundance
top <- circus.df  %>% slice_max(total, n= 9) 
top <- circus.df  %>% slice_max(total, n= 15) 

# make another data frame with the others taxa
no_top <- circus.df %>% anti_join(top) %>% # to make a complete dataframe use bind_rows()
  summarise(.,
            across(where(is.numeric), sum),
            across(where(is.character), ~"others"))

# add to the first dataframe
circus.df.plot <- bind_rows(top,no_top) %>%
  # map function to map each column to the function you mention.
  # map_at(c("a", "b", "c"), ~./sum(.)) %>% 
  map_if(is.numeric, ~./sum(.)) %>%
  as.data.frame()  %>%
  select(-total) %>%
  column_to_rownames({{Rank1}})

# if there are less than 20 taxa in the data to plot
circus.df.plot <- circus.df %>%
  # map function to map each column to the function you mention.
  # map_at(c("a", "b", "c"), ~./sum(.)) %>% 
  map_if(is.numeric, ~./sum(.)) %>%
  as.data.frame()  %>%
  select(-total) %>%
  column_to_rownames({{Rank1}})
```


```{r circus plot 2 - table}
# save a table
kingdom <- ps.melt1 %>% select(Kingdom) %>% pull %>% unique
to_save <- circus.df.plot %>%
  rownames_to_column({{Rank1}}) %>%
          mutate(across(is.numeric, ~. *100),
          across(is.numeric, ~ round(.,2)))
write.csv(to_save, 
          paste0('C:/Users/rrocha/Documents/R/phd/2018/output/plots/all/composition/circus/circus_',{{kingdom}},"_",{{Rank}},"_",taxa.to.filter,"_filtered_table",".csv"))
```


```{r circus plotting 2}
# plotting
# set colors
#library(RColorBrewer)
colourCount <- circus.df.plot %>% row.names() %>% n_distinct()
# colourCount <-length(unique(ps.glom[,length(ps.glom)]))
getPalette <- colorRampPalette(brewer.pal(12, "Paired"))
grid.col <- circus.df.plot %>% rownames_to_column('taxa') %>%
  add_column(col = getPalette(colourCount)) %>%
  pull(col, taxa)
grid.col <- c(grid.col, cols_hab_layer)

# prep
mat <- as.matrix(circus.df.plot)

# clean the plot area
circos.clear()
par(xpd=TRUE) # to enable things to be drawn outside the plot region. 

# Basic circos graphic parameters
circos.par(start.degree = 115, gap.after = c(rep(2, nrow(mat)-1), 10, rep(2, ncol(mat)-1), 10))

# script
chordDiagram(mat, grid.col = grid.col, directional = 1, direction.type = c("diffHeight", "arrows"),
             link.arr.type = "big.arrow", annotationTrack = c("grid"), scale = FALSE)

# axis and lables 
circos.track(track.index = 1, panel.fun = function(x, y) {
  xlim = get.cell.meta.data("xlim")
  ylim = get.cell.meta.data("ylim")
  sector.name = get.cell.meta.data("sector.index")
  xplot = get.cell.meta.data("xplot")
  circos.points(x = x, y = y, pch = "·")
  #plot labels
  circos.text(mean(xlim), y= 3 ,sector.name, facing = "reverse.clockwise", niceFacing = TRUE, adj = aa, cex = 1)
  #plot axis
  circos.axis(labels.cex=0.6, direction = "outside")
 
}, bg.border = NA) # here set bg.border to NA is important

circos.clear()

par(mfrow=c(1,1)) # reset plotting area
```


### Bar plot

#### Site
##### Phylum
```{r phylum}
# https://github.com/adriaaulaICM/bbmo_niche_sea/blob/master/src/figures/abund_prevalence_viz.R
# create file directory
# dir.create(paste0(folder_path,"plots/barplots_18s"))

# all
bar_plot_phylum_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "site")

bar_plot_phylum_all <- bar_plot_phylum_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_phylum_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_all_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_phylum_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "site")

bar_plot_phylum_phyto <- bar_plot_phylum_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_phylum_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_phyto_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_phylum_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "site")

bar_plot_phylum_other <- bar_plot_phylum_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_phylum_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_other_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Class
```{r class}
# all
bar_plot_class_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "site")

bar_plot_class_all <- bar_plot_class_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_class_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_all_class.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_class_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "site")

bar_plot_class_phyto <- bar_plot_class_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_class_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_phyto_class.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_class_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "site")

bar_plot_class_other <- bar_plot_class_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_class_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_other_class.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Order
```{r order}
# all
bar_plot_order_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "site")

bar_plot_order_all <- bar_plot_order_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_order_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_all_order.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_order_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "site")

bar_plot_order_phyto <- bar_plot_order_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_order_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_phyto_order.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_order_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "site")

bar_plot_order_other <- bar_plot_order_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_order_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_other_order.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Family
```{r family}
# all
bar_plot_family_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "site")

bar_plot_family_all <- bar_plot_family_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_family_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_all_family.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_family_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "site")

bar_plot_family_phyto <- bar_plot_family_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_family_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_phyto_family.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_family_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "site")

bar_plot_family_other <- bar_plot_family_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_family_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/bar_plot_18s_other_family.tiff"),
       width = 16, height = 10, compression = "lzw")

```

#### Depth
##### Phylum
```{r phylum}
# https://github.com/adriaaulaICM/bbmo_niche_sea/blob/master/src/figures/abund_prevalence_viz.R
# create file directory
# dir.create(paste0(folder_path,"plots/barplots_18s"))
pseq.euk.list[[1]] %>% get_tibble("sam_data") 
# all
bar_plot_phylum_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "Depth1")

bar_plot_phylum_all <- bar_plot_phylum_all + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_phylum_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_all_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_phylum_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "Depth1")

bar_plot_phylum_phyto <- bar_plot_phylum_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_phylum_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_phyto_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_phylum_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Phylum", grouping_column = "Depth1")

bar_plot_phylum_other <- bar_plot_phylum_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_phylum_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_other_phylum.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Class
```{r class}
# all
bar_plot_class_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "Depth1")

bar_plot_class_all <- bar_plot_class_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_class_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_all_class.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_class_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "Depth1")

bar_plot_class_phyto <- bar_plot_class_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_class_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_phyto_class.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_class_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Class", grouping_column = "Depth1")

bar_plot_class_other <- bar_plot_class_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_class_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_other_class.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Order
```{r order}
# all
bar_plot_order_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "Depth1")

bar_plot_order_all <- bar_plot_order_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_order_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_all_order.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_order_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "Depth1")

bar_plot_order_phyto <- bar_plot_order_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_order_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_phyto_order.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_order_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Order", grouping_column = "Depth1")

bar_plot_order_other <- bar_plot_order_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_order_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_other_order.tiff"),
       width = 16, height = 10, compression = "lzw")

```

##### Family
```{r family}
# all
bar_plot_family_all <- pseq.euk.list[[1]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "Depth1")

bar_plot_family_all <- bar_plot_family_all  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))

ggsave(bar_plot_family_all, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_all_family.tiff"),
       width = 16, height = 10, compression = "lzw")


# Phytoplankton
bar_plot_family_phyto <- pseq.euk.list[[2]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "Depth1")

bar_plot_family_phyto <- bar_plot_family_phyto  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_family_phyto, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_phyto_family.tiff"),
       width = 16, height = 10, compression = "lzw")


# other-protists
bar_plot_family_other <- pseq.euk.list[[3]] %>% 
  microbiome::transform("compositional") %>% 
  bar_plot_taxa_group(taxo = "Family", grouping_column = "Depth1")

bar_plot_family_other <- bar_plot_family_other  + 
                          theme(legend.position = "bottom",
                                legend.text = element_text(size = 12))


ggsave(bar_plot_family_other, 
       filename = paste0(folder_path, "/plots/barplots_18s/depth_bar_plot_18s_other_family.tiff"),
       width = 16, height = 10, compression = "lzw")

```

### PCA

```{r Exploratory PCA}
load("pseq.euk.list.clr.RData") # clr transformed data
pseq.euk.list.clr

pca_pseq <- function(x,...){
pseq <- x
clr_otu <- otu_table(pseq)
# CoDaSeq::codaSeq.outlier(t(clr_otu))
ord_meta <- pseq %>% get_tibble("sam_data", column_id = "sample_id")
King <- unique(data.frame(tax_table(pseq)[,1]))
Y <- as.data.frame(t(clr_otu)) 

pca_fit <- Y %>% 
  scale %>% 
  prcomp() 

var <- pca_fit %>% 
  tidy(matrix = "eigenvalues")

p.pca <- pca_fit %>%
  augment() %>%
  rename("sample_id" = 1) %>%
  left_join(ord_meta) %>%
  rename_with(~str_replace(.,".fitted","")) %>%
  ggplot(aes(PC1, PC2, color = Depth)) + 
  geom_point(aes(color = Depth, shape = site), alpha = 0.7, size =3) + 
  #geom_text(check_overlap = TRUE, hjust = 'inward', family = 'IBM Plex Sans') +
  geom_hline(yintercept=0, linetype="dashed", alpha = 0.3) +
  geom_vline(xintercept=0, linetype="dashed", alpha = 0.3) +
  labs(x = paste("PC1", round(var$percent[1]*100,2), "%"),
       y = paste("PC2", round(var$percent[2]*100,2), "%")) +
  # scale_colour_viridis_d(option = "plasma") 
  scale_color_manual(labels = unique(ord_meta$Depth1),values = color_depth) + 
  scale_shape_manual(values=c(15, 18, 17)) +
  ggtitle(King)

return(p.pca)
}

# clr transformed
clr.pca <- pseq.euk.list.clr %>% map(pca_pseq)
clr.pca.18s <- clr.pca %>% keep(names(.) %in% c('Phytoplankton', "Other_protists"))


clr.pca.all.18s <- wrap_plots(clr.pca.18s) + 
  plot_layout(guides='collect')  &
  theme(legend.position='bottom')

ggsave(paste0(folder_path, "plots/community_18s/pca.18s_phyto_other.tiff"),clr.pca.all.18s, width = 14, height = 6.5, compression = "lzw")

```

### Hierarquical Clustering

```{r EDA hcluster}
# hierarchical clustering
# Aitchison distance = Euclidean distance of clr-transformed compositions
# generate the distance matrix
taxdf <- pseq.euk.list.clr[[3]] %>% 
  get_tibble("tax_table") %>% 
  distinct(Phylum) %>% pull

#  pseq <- pseq.func %>% transform("clr")
df <- pseq.euk.list.clr[[3]] %>% abundances()
dd <- dist(t(df), method="euclidian")
# cluster the data
hc <- hclust(dd, method="ward.D2")
# str(hc)
# Cluster plot
plot(hc, cex=0.6)
# generate colors
library(scales)
n1 <- 5                                        # Amount of default colors
hex_codes1 <- hue_pal()(n1)                             # Identify hex codes
#show_col(hex_codes1)

# plot using the optimal number of clusters found in the k-mean cluster
clust1 <- factoextra::fviz_dend(hc, cex = 0.8, k= 6, palette = c("darkblue","#A3A500","darkgreen","#00B0F6","#dd00f6","darkred"), 
                    rect = TRUE, rect_border = c("darkblue","#A3A500","darkgreen","#00B0F6","#dd00f6","darkred"), rect_fill = TRUE, 
                    main = "Dendrogram - ward.D2",
                    xlab = "Samples", ylab = "Distance", sub = "") +
  theme_Publication_1() +
  theme(panel.border = element_blank())

```

### db-RDA

```{r}
# dir.create(paste0(folder_path,"data/community_18s/"))
```


ordistep, VIF and envfit used to select the env variables
```{r RDA}
# prepare the data
rda.pseq <- function(pseq, title){
  set.seed(123)
tax_df <- pseq %>% get_tibble("tax_table", column_id = "ASV") 
# King <- tax_df %>% distinct(Kingdom) %>% pull
clr_asv <- pseq %>% 
    get_tibble("otu_table", column_id = "ASV") %>% 
    column_to_rownames('ASV')
meta <- pseq %>% 
    get_tibble("sam_data", column_id = "sample_id")
ord_meta <- meta %>%
    select(sample_id, NH4, NOX, turbidity, oxygen) %>% 
    column_to_rownames("sample_id")
  
# cond <- apply(ord_meta, 2, function(x) any(is.na(x))) # check for NAs
X <- ord_meta %>% select(where(~is.numeric(.) && all(!is.na(.))))
X <- vegan::decostand(X, method='standardize')
Y <- as.data.frame(t(clr_asv)) 
  
# RDA
res <- rda(Y ~ ., data=X)
# plot(res, xlab=QsRutils::ord_labels(res)[1], ylab=QsRutils::ord_labels(res)[2])
  
# dbRDA 
# note the result is the same as for RDA when using euclidean distances
# res <- capscale(Y ~ ., data=X, distance='euclidean')
# plot(res, xlab=QsRutils::ord_labels(res)[1], ylab=QsRutils::ord_labels(res)[2])

# select only the variables that are explaining variation efficiently.

# calculate variance inflation factors (VIF)
# variables with scores >10 are redundant
# sort(vif.cca(res))
  
# Fitting environmental vectors/factors onto an ordination 
# calculate fit
bio.fit <- envfit(Y ~ ., data=X)
  
# Stepwise selection (ordistep)
# set up full and null models for 'ordistep'
# full model
rda1 <- rda(Y ~ ., data=X)
# intercept-only (null) model
rda0 <- rda(Y ~ 1, data=X)
  
# perform forward and backward selection of explanatory variables
# output not shown
step.env <- ordistep(rda0, Pin = 0.01, scope=formula(rda1), direction='both')
  
# look at the significant variables 
#step.env$anova
# For the sake of argument, let’s include the variables
# with low VIF scores along with those identified by ordistep.
# code to get variable names from 'ordistep' and 'envfit' results
# make a dataframe with pvalues and envpars
vars.df <- data.frame(step.env$anova) %>% 
    rownames_to_column("env_par") %>%
    mutate(env_par = str_trim(str_replace(env_par, "[+]", ""))) %>% 
    # mutate(env_par = str_replace(env_par, "\"", "")) %>% 
    rename(p_value.ordistep = Pr..F.) %>% 
    select(env_par,p_value.ordistep) %>% 
    full_join(data.frame(bio.fit$vectors$pvals) %>% filter(. <=0.01) %>%
                rename(p_value.env.fit = bio.fit.vectors.pvals) %>%
                rownames_to_column("env_par")
    ) %>%
    full_join(as.data.frame(vif.cca(res)) %>% filter(. <=10) %>%
                rename(vif.score = 'vif.cca(res)') %>%
                rownames_to_column("env_par")
    ) %>%
    full_join(bio.fit[["vectors"]][["r"]] %>% data.frame() %>% 
                rename(r_envfit = '.') %>%
                rownames_to_column("env_par") %>%
                mutate(r_envfit = round(r_envfit,2))
    )

# get the variables to use in the next steps of the analysis
vars <- vars.df %>% 
  filter(p_value.ordistep < 0.01 & p_value.env.fit < 0.01) %>% 
  pull(env_par)

# write_csv(vars.df, paste0(folder_path, "data/db-RDA_bacteria.csv"))  
#vars 
# select variables to keep from table 'Y'
X1 <- X %>% select(matches(vars))
#str(X1)
  
# RDA
res <- rda(Y ~ ., data=X1)
  
  # summary of the results
  # summary(res, display=NULL)
  # permutation test of statistical significance
  # anova(res)
  
  # plotting
  # load library
  
# set up dataframes for plotting the results
sit <- meta %>%
    left_join(vegan::scores(res, 
                            display='sites', choices=c(1:2)) %>%
                as.data.frame() %>%
                rownames_to_column("sample_id")) %>%
  column_to_rownames("sample_id")

spp <- tax_df %>% 
    left_join(vegan::scores(res, display='species') %>% 
                data.frame() %>% 
                rownames_to_column('ASV')) %>%
                column_to_rownames("ASV")
vec <- vegan::scores(res, display='bp') %>% 
    data.frame() %>%
    rownames_to_column("terms") %>%
    mutate(terms = case_when(terms == "oxygen" ~ "oxygen, PAR and fluorescence",
                           terms == "NH4" ~ "NH4 and NO2",
                           terms == "NOX" ~ "NOX, conductivity, salinity, \ntemperature and PO4",
                           TRUE ~ as.character(terms))) 
  
  # use these to adjust length of arrows and position of arrow labels
  adj.vec <- 2
  adj.txt <- 2.5
  arrow_style <- arrow(length = unit(.05, "inches"),
                       type = "closed")
  
  # 'site' ordination
p1 <- sit %>%
    rownames_to_column("sample_id") %>%
    ggplot(aes(x=RDA1, y=RDA2, label = sample_id)) +
    geom_point(aes(color = Depth, shape = site) , size= 3,alpha = 0.8) +
    geom_segment(data = vec,
                 inherit.aes=F,
                 mapping=aes(x=0, y=0, 
                             xend=adj.vec*RDA1, 
                             yend=adj.vec*RDA2),
                 arrow = arrow_style,
                 alpha = 0.4) +
      ggrepel::geom_text_repel(data=vec,
              mapping=aes(x=adj.txt*RDA1, y=adj.txt*RDA2,
                          label=terms), 
              hjust = 0,
              vjust = 0.65,
              size = 4.2,
              alpha = 0.9,
              color = '#0A234E') +
  #  ggtitle(paste0('dbRDA: ',King)) +
    scale_color_manual(labels = unique(meta$Depth1),values = color_depth) + 
    scale_shape_manual(values=c(15, 18, 17)) +
    geom_hline(yintercept=0, linetype="dashed", alpha = 0.3) +
    geom_vline(xintercept=0, linetype="dashed", alpha = 0.3) +
    #labs(colour = GroupBy)+ #another way to set the labels, in this case, for the colour legend
    labs(x = QsRutils::ord_labels(res)[1],
         y = QsRutils::ord_labels(res)[2],
         title = title) +
    theme(legend.position = 'bottom') 


return(list(p1, vars.df))

}

rda_results <- pseq.euk.list.clr %>% map2(.,names(.), ~rda.pseq(.x, .y))

# save env drivers in a csv table
rda_results %>% 
  map(., 2) %>% 
  map2(names(.), ~add_row(.x, env_par = .y, .before = 1)) %>%
  bind_rows() %>%
  write_csv(paste0(folder_path, "data/community_18s/db-RDA_18s.csv")) 

# plot all together
plots_rda_18s <- rda_results %>% 
  map(., 1) %>%
  keep(names(.) %in% c("Phytoplankton", "Other_protists" ))
  

wrapped <- wrap_plots(plots_rda_18s) & theme(legend.position = "bottom")
wrapped <- wrapped + plot_layout(guides = "collect")

# 14.1 x 6.39 in
# 12.4 x 6.93 in
ggsave(wrapped, filename = paste0(folder_path, "plots/community_18s/db-RDA_18s_phyto_other.tiff"), compression = "lzw")

```


### Mantel test

A significant Mantel test will tell you that the distances between samples in one matrix are correlated with the distances between samples in the other matrix. Therefore, as the distance between samples increases with respect to one matrix, the distances between the same samples also increases in the other matrix.

Here we will analyze if the differences in community composition between samples are correlated (or co-vary) with the differences in environmental variables (i.e., environment is selecting) or physical distance (i.e., distance decay pattern - dispersal limitation) between samples.

Using simple Mantel test, we will test if cumulative environmental factors (combination of all parameters) and distance (spatial gradient) from the cage are correlated with microbial community. Next, we will split the dataset by depth to verify this correlation in each water layer.

Good results summary
<https://doi.org/10.1016/j.marpolbul.2017.04.059>

#### run the models
```{r mantel}
mantel_func <- function(pseq, title){

asv_clr <- pseq %>% 
  get_tibble("otu_table", column_id = "ASV") %>% 
  column_to_rownames("ASV")
meta.mantel <-  pseq %>% 
  get_tibble("sam_data", column_id = "sample_id") %>% 
  column_to_rownames("sample_id") 

#abundance data frame
abund = data.frame(t(asv_clr)) # SAMPLES as ROWS !!!!!
#abundance data frame - bray curtis dissimilarity
dist.abund = dist(abund, method = "euclidean")

#make subset and remove atutocorrelated variables
env = meta.mantel %>% 
  dplyr::select(NH4, NOX, oxygen, turbidity)
#scale data 
scale.env = scale(env, center = TRUE, scale = TRUE)
#create distance matrix of scaled data
dist.env = dist(scale.env, method = "euclidean")

#longitude and latitude 
geo = meta.mantel %>% select(long, lat)
#geographic data frame - haversine distance 
d.geo = geosphere::distm(geo, fun = geosphere::distHaversine)
dist.geo = as.dist(d.geo)

# run mantel test 
# abundance vs environmental
abund_env = mantel(dist.abund, dist.env, method = "spearman", permutations = 9999, na.rm = TRUE)
abund_env
#abundance vs geographic 
abund_geo  = mantel(dist.abund, dist.geo, method = "spearman", permutations = 9999, na.rm = TRUE)
abund_geo
# environmental vs geographic
env_geo  = mantel(dist.env, dist.geo, method = "spearman", permutations = 9999, na.rm = TRUE)
env_geo

# Partial Mantel tests
# geo distance controlled
# We will do that while removing any possible spatial autocorrelation.
abund_partial_envgeo = mantel.partial(dist.abund, dist.env, dist.geo, method = "spearman", permutations = 9999, na.rm = TRUE)
abund_partial_envgeo
# env variability controlled
abund_partial_geoenv  = mantel.partial(dist.abund,  dist.geo, dist.env,method = "spearman", permutations = 9999, na.rm = TRUE)
abund_partial_geoenv   

df <- tibble(test = c('abund_env', 'abund_geo', 'env_geo', "partial_ab_env_by_geo", "partial_ab_geo_by_env"),
          rho = c(abund_env$statistic, abund_geo$statistic, env_geo$statistic,abund_partial_envgeo$statistic, abund_partial_geoenv$statistic),
          p_value = c(abund_env$signif, abund_geo$signif, env_geo$signif, abund_partial_envgeo$signif,abund_partial_geoenv$signif))

# Pairwise scatter plots
aa = as.vector(dist.abund)
tt = as.vector(dist.env)
gg =  as.vector(dist.geo)

#new data frame with vectorized distance matrices
mat = data.frame(aa,tt,gg)

#abundance vs env paramters
lm = lm(aa~tt, data = mat)
summary(lm)
my.formula <- aa~tt

ab_env <- ggplot(mat, aes(y = aa, x = tt)) + 
  geom_point(size = 2.5, alpha = 0.3) + 
  labs(x = "Difference in Env Factors", y = "Euclidean Distance") + 
  geom_smooth(method = 'lm', color = "orange", alpha = 0.3, se = F)  +
  # show rho and p.value
  ggpubr::stat_cor(method = "spearman",
                   cor.coef.name = "rho",
                   aes(label = paste(..r.label.., ..p.label.., sep = "~`,`~"))) +
  ggtitle(glue::glue("{title}"))

#abundance vs geo dist
lm = lm(aa~gg, data = mat)
summary(lm)
my.formula <- aa~gg

#new data frame with vectorized distance matrices
ab_geo <- ggplot(mat, aes(y = aa, x = gg/1000)) + 
  geom_point(size = 2.5, alpha = 0.3) + 
  labs(x = "Physical Distance (km)", y = "Euclidean Distance") + 
  geom_smooth(method = 'lm', color = "orange", alpha = 0.3, se = F)  +
  # show rho and p.value
  ggpubr::stat_cor(method = "spearman",
                   cor.coef.name = "rho",
                   aes(label = paste(..r.label.., ..p.label.., sep = "~`,`~"))) +
  ggtitle(glue::glue("{title}"))


return(list(df, ab_env, ab_geo))

}

# all depths
mantel1 <- pseq.euk.list.clr %>% map2(., names(.), ~mantel_func(.x,.y))

# DP1
mantel2 <- pseq.euk.list.clr %>% 
  map(., ~subset_samples(., Depth == "DP1")) %>% 
  map2(., names(.), ~mantel_func(.x, paste0(.y," - Surface")))

# DP2
mantel3 <- pseq.euk.list.clr %>% 
  map(., ~subset_samples(., Depth == "DP2")) %>% 
  map2(., names(.), ~mantel_func(.x, paste0(.y," - Middle layers")))

# DP3
mantel4 <- pseq.euk.list.clr %>% 
  map(., ~subset_samples(., Depth == "DP3")) %>% 
  map2(., names(.), ~mantel_func(.x, paste0(.y," - Bottom waters")))
```

#### make the datastes
```{r mantel}
df_mantel_al_euk <- bind_rows(mantel1[[1]][[1]] %>% add_row(test = "all",.before = 1), 
          mantel2[[1]][[1]] %>% add_row(test = "surface",.before = 1), 
          mantel3[[1]][[1]] %>% add_row(test = "middle",.before = 1), 
          mantel4[[1]][[1]] %>% add_row(test = "bottom",.before = 1))

write_csv(df_mantel_al_euk, paste0(folder_path,"data/community_18s/mantel_all.csv"))

# all depths
df_mantel_all_euk <- map(mantel1, 1) %>% 
  map2(.,names(.) ,~add_row(.,test = paste(.y), .before = 1)) %>% 
  bind_rows()

write_csv(df_mantel_all_euk, paste0(folder_path,"data/community_18s/mantel_all.csv"))

# surface
df_mantel_surf_euk <- map(mantel2, 1) %>% 
  map2(.,names(.) ,~add_row(.,test = paste(.y), .before = 1)) %>% 
  bind_rows()

write_csv(df_mantel_surf_euk, paste0(folder_path,"data/community_18s/mantel_surface.csv"))

# middle
df_mantel_mid_euk <- map(mantel3, 1) %>% 
  map2(.,names(.) ,~add_row(.,test = paste(.y), .before = 1)) %>% 
  bind_rows()

write_csv(df_mantel_mid_euk, paste0(folder_path,"data/community_18s/mantel_middle.csv"))

# depth
df_mantel_bottom_euk <- map(mantel4, 1) %>% 
  map2(.,names(.) ,~add_row(.,test = paste(.y), .before = 1)) %>% 
  bind_rows()

write_csv(df_mantel_bottom_euk, paste0(folder_path,"data/community_18s/mantel_bottom.csv"))
```

Results:
Further analysis were performed using Mantel and partial Mantel tests to explorer the co-variation of the cumulative environmental factors, spatial gradient and eukaryotic communities.  ....
However, the partial Mantel tests showed that the correlation considering all depths tohgether became weak (| ρ | < 0.5), indicating a partly dependence on depth (ρ = 0.416, P = 0.024, Table 2) and geographical distance (ρ = 0.491, P = 0.014, Table S2).
permanceu significante quando se cntrolou a distancia, indicadon que a variacao nao eh dependendte da distance. qd se dividiu por profundiade as correlacoes se tornaram fracas mostrando uma forte dependencia co ma profundiade. (ρ = xxxx, P = xxxx, Table 2)

```{r mantel}
# mantel_plots <- (mantel1[[3]] + theme(axis.title.x = element_blank()) + mantel2[[3]] + theme(axis.title.y = element_blank(), axis.title.x = element_blank())) / (mantel3[[3]] + mantel4[[3]] + theme(axis.title.y = element_blank()))
# 
# mantel_plots1 <- (mantel1[[2]] + theme(axis.title.x = element_blank()) + mantel2[[2]] + theme(axis.title.y = element_blank(), axis.title.x = element_blank())) / (mantel3[[2]] + mantel4[[2]] + theme(axis.title.y = element_blank()))
# 
# # ylab <- mantel1$labels$y
# # xlab <- mantel3$labels$x
# # mantel1$labels$y <- mantel3$labels$y <- " "
# # mantel3$labels$x <- mantel4$labels$x <- " "
# 
# # mantel_plots <- (mantel1 + mantel2) / (mantel3 + mantel4)
# # grid::grid.draw(grid::textGrob(ylab, x = 0.02, rot = 90))
# # grid::grid.draw(grid::textGrob(xlab, y = 0.02, rot = 360))
# 
# ggsave(paste0(folder_path,"/plots/community/mantel_bac_distgeo.tiff"), mantel_plots, compression = "lzw")
# 
# ggsave(paste0(folder_path,"/plots/community/mantel_bac_env.tiff"), mantel_plots1, compression = "lzw")
# 
# 
# # Mantel test (mantel_test.R)
# # https://jkzorz.github.io/2019/07/08/mantel-test.html


```



