---
title: "feature_selection"
author: "Ricardo Silva"
date: "02/09/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval=FALSE, 
                      message=FALSE, 
                      warning=FALSE)
```

```{r CLEAR EVERYTHING, eval=FALSE, include=FALSE}
# unload all non-base packages
lapply(names(sessionInfo()$otherPkgs), function(pkgs)
  detach(
    paste0('package:', pkgs),
    character.only = T,
    unload = T,
    force = T
))

rm(list=ls())
```

# Packages and initial steps

```{r load libraries, message=FALSE, warning=FALSE, include=FALSE}
# load the packages
# main packs to start with
pcks <- c('phyloseq', 'microbiome','microbiomeutilities' , "patchwork","ggtern","scales","ggforce",'tidyverse')

if(sum(as.numeric(!pcks %in% installed.packages())) != 0){
  installation <- pcks[!pcks %in% installed.packages()]
  for(i in 1:length(installation)) {
    install.packages(installation, dependencies = T)
    break()}
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
} else {
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
}

rm(pcks)
```

```{r packages and functions}
# set working directory and seed
setwd("C:/Users/rrocha/Documents/R/phd/2019/data")
folder_path <- "C:/Users/rrocha/Documents/R/phd/2019/output/"

source('~/R/phd/2019/script/my_functions.R')
source('~/R/phd/2019/script/theme_publication.R')
theme_set(theme_Publication_3())

# set colors
color_distance <- c("D0" = "deepskyblue2", 'D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# color_distance1 <- c('D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
color_depth <- viridis::viridis(n=3)
lease_control <- friendly_cols

# check the working directory
current_directory <- getwd()
paste("Working in directory:", current_directory)
```

# Aims

Here, we used three approaches (Deseq2, ALDEx2, and  WGNCA) to select the ASVs most correlated to lease1, lease2 or control sites

Lease 1 = B (T1 and T2)
Lease 2 = G (T3, T4 and T5)

# Data

## meta and pseq objects

```{r load data}
# load object in RData format
# 18S
# load("pseq.euk.list.RData")
# pseq.euk.list
load("pseq.euk.list.filt.RData") # filtered data
pseq.euk.list.filt
# load("pseq.euk.list.clr.RData") # clr transformed data
# pseq.euk.list.clr

# meta
meta <- read_csv(paste0(getwd(),"/meta_impute.csv")) 

dir.create(paste0(folder_path, "data/feature_selection_18s"))
dir.create(paste0(folder_path, "plots/feature_selection_18s"))

```

# Feature selection

## ALDEx2

Get the OTUs with an effect size greater than |1.5|
for example, in the condition conds.aldex.1.2 <- c(cond1, cond2) = cond1 vs cond2
ASVs with a higher relative abundance in cond2 will have positive effect
values, OTUs higher in cond1 will have negative effect values
we can use Wilcoxon test results: wi.eBH (FDR), or raw p values (wi.ep, we.ep)
from either test

<https://github.com/ggloor/CoDa_microbiome_tutorial/wiki/>

```{r aldex function}
# function do run ALDEx2 between lease and control by each depth
aldex_fun <- function(physeq, filter_statement){

  s <- physeq %>% get_tibble("sam_data") %>%
    filter(eval(rlang::parse_expr(filter_statement)))  %>% 
    pull(column_id)
  pseq.filt <- prune_samples(s, physeq)
  # removing species not seen > 1 samples (prevalence > 1)
  pseq = filter_taxa(pseq.filt, function(x) sum(x > 0) > 1, TRUE)
  
#  pseq <- filter_function(physeq, var1, var2, var3)
#  pseq <- p.filt %>% subset_samples(site %in% c("lease_2", "lease_1"))
  
  asv_table <- pseq %>% get_tibble("otu_table") %>%
    column_to_rownames("column_id")
  TAXdf <- pseq %>% get_tibble("tax_table") %>%
    column_to_rownames("column_id")
  
  # make a vector containing the two names of the conditions
  # in the same order as in the column names
  ## To know how many samples per depth
  control <-  length(str_subset(colnames(asv_table), "C")) 
  lease <- length(str_subset(colnames(asv_table), "T"))
  
  # make a data set with the conditions in right order
  d.aldex <- asv_table %>% dplyr::select(starts_with("C") | starts_with("T") )
  
  # conditions (clusters)
  conds.aldex <- c(rep("control", control), rep("lease", lease))
  
  ## CHECK this vector in your console by typing: 
  conds.aldex
  colnames(d.aldex)
  # This should match the order and number of columns you have: 
  waldo::compare(colnames(d.aldex), conds.aldex)
  
  # Calculate the Kruskal-Wallis test and glm ANOVA statistics
  ## x.w <- ALDEx2::aldex.kw(x)
  x.all <- ALDEx2::aldex(d.aldex , conds.aldex, mc.samples=128, test="t", effect=TRUE, include.sample.summary=FALSE, verbose=FALSE)
  
  # generate a datasetc with all OTUs 
  x.all.asv <- x.all %>% 
    rownames_to_column(var = "ASV") %>%
    left_join(TAXdf %>% rownames_to_column('ASV'))
  
  return(x.all.asv)
}

```

```{r}
# preparare the interaction before suing map
# depth <- c("DP1", "DP2", "DP3")
# sites <- c("Control", "lease_1", "lease_2") 
# king <- names(pseq.euk.list.filt)
# 
# exp <- cross_df(list(x = depth, y = sites, z= sites)) %>% 
#   filter(y != z & z != "Control" & y !="lease_2") %>%
#   mutate(expression = glue::glue("site %in% c('{y}', '{z}') & Depth == '{x}'")) %>%
#   pull(expression)

# preparare the interaction before suing map
depth <- c("DP1", "DP2", "DP3")
control <- c("Control") 
leases <- c("lease_1", "lease_2")
king <- names(pseq.euk.list.filt)

exp <- crossing(x = depth, y = control, z=leases) %>% 
  mutate(expression = glue::glue("site %in% c('{y}', '{z}') & Depth == '{x}'"),
         name.list = glue::glue("{x} - {y} vs {z}"))


```

## Control vs leases 1 and 2

### 18S

```{r ALDEx2 - 18s control vs leases}
# Eukaryotes
aldex.euk <- map(exp$expression, ~aldex_fun(pseq.euk.list.filt[[1]],.x)) %>%
  set_names(exp$name.list)

aldex.euk.filt <- aldex.euk %>%
  map2(.,names(.), ~mutate(., contrast = paste(.y), .before = 1)) %>%
  map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
#  map2(.,names(.), ~add_row(., ASV = .y, .before = 1)) %>% 
  bind_rows()  %>%
  mutate(method = "ALDEx2", .before = 1)



# Phytoplankton
aldex.phyto <- map(exp$expression, ~aldex_fun(pseq.euk.list.filt[[2]],.x)) %>%
  set_names(exp$name.list)

aldex.phyto.filt <- aldex.phyto %>%
  map2(.,names(.), ~mutate(., contrast = paste(.y), .before = 1)) %>%
  map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
#  map2(.,names(.), ~add_row(., ASV = .y, .before = 1)) %>% 
  bind_rows()  %>%
  mutate(method = "ALDEx2", .before = 1)



# other protists
aldex.not.phyto <- map(exp$expression, ~aldex_fun(pseq.euk.list.filt[[3]],.x)) %>%
  set_names(exp$name.list)

aldex.not.phyto.filt <- aldex.not.phyto %>%
  map2(.,names(.), ~mutate(., contrast = paste(.y), .before = 1)) %>%
  map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
#  map2(.,names(.), ~add_row(., ASV = .y, .before = 1)) %>% 
  bind_rows() %>%
  mutate(method = "ALDEx2", .before = 1)


```

## lease 1 vs lease 2

```{r aldex function}
# function to filter samples
filter_function <- function(physeq, var1, var2, var3) {
  s <- physeq %>%
    get_tibble("sam_data") %>%
    filter(site == !!(var1) | site == !!(var2)) %>%
    filter(Depth == !!(var3)) %>%
    pull(column_id)
  pseq.subset <- prune_samples(s, physeq)
    return(pseq.subset)
}

# function do run ALDEx2 between lease and control by each depth
aldex_fun2 <- function(physeq, var1, var2, var3){

  pseq.filt <- filter_function(physeq, var1, var2, var3)
    # removing species not seen > 1 samples (prevalence > 1)
  pseq = filter_taxa(pseq.filt, function(x) sum(x > 0) > 1, TRUE)
  #  pseq <- p.filt %>% subset_samples(site %in% c("lease_2", "lease_1"))
  TAXdf <- pseq %>% get_tibble("tax_table") %>%
    column_to_rownames("column_id")
  
  # make a vector containing the two names of the conditions
  # in the same order as in the column names
  ## To know how many samples per depth
  cond1 <- pseq %>% get_tibble("sam_data") %>% filter(site == var1) %>% count()
  cond2 <- pseq %>% get_tibble("sam_data") %>% filter(site == var2) %>% count()

  # make a data set with the conditions in right order
  # d.aldex <- asv_table %>% dplyr::select(starts_with("C") | starts_with("T") )
  
  d.aldex <- pseq %>% get_tibble("otu_table") %>%
    column_to_rownames("column_id") %>% 
    select(order(colnames(.)))
  
  # conditions (clusters)
  conds.aldex <- c(rep(glue::glue("{var1}"), cond1), rep(glue::glue("{var2}"), cond2))
  
  ## CHECK this vector in your console by typing: 
  # return(conds.aldex)
  # return(colnames(d.aldex))
  # This should match the order and number of columns you have: 
  # return(waldo::compare(colnames(d.aldex), conds.aldex))
  
  # Calculate the Kruskal-Wallis test and glm ANOVA statistics
  ## x.w <- ALDEx2::aldex.kw(x)
  x.all <- ALDEx2::aldex(d.aldex , conds.aldex, mc.samples=128, test="t", effect=TRUE, include.sample.summary=FALSE, verbose=FALSE)
  
  # generate a datasetc with all OTUs with effect size > |0.5|
  x.all.asv <- x.all %>% 
    rownames_to_column(var = "ASV") %>%
    left_join(TAXdf %>% rownames_to_column('ASV'))
  
  return(x.all.asv)
}


```

### 16S and 18S

```{r ALDEx2 - 16s control vs leases}
# DP1
aldex.leases.dp1 <- pseq.euk.list.filt %>% 
  set_names() %>%
  map(.,~aldex_fun2(., "lease_1", "lease_2", "DP1"))

aldex.leases.dp1.filt <- aldex.leases.dp1  %>%
    map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
  bind_rows()  %>%
  mutate(method = "ALDEx2", .before = 1,
         contrast = "DP1 - lease_1 vs lease_2")


# DP2
aldex.leases.dp2 <- pseq.euk.list.filt %>% 
  set_names() %>%
  map(.,~aldex_fun2(., "lease_1", "lease_2", "DP2"))

aldex.leases.dp2.filt <- aldex.leases.dp2  %>%
    map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
  bind_rows()  %>%
  mutate(method = "ALDEx2", .before = 1,
         contrast = "DP2 - lease_1 vs lease_2")


# DP3
aldex.leases.dp3 <- pseq.euk.list.filt %>% 
  set_names() %>% 
  map(.,~aldex_fun2(., "lease_1", "lease_2", "DP3"))


aldex.leases.dp3.filt <- aldex.leases.dp3  %>%
    map(.,~filter(., effect >= 1 | effect <= -1) %>% arrange(effect)) %>%
  bind_rows()  %>%
  mutate(method = "ALDEx2", .before = 1,
         contrast = "DP3 - lease_1 vs lease_2")

# bind_all
all_aldex.leases <- bind_rows(aldex.leases.dp1.filt,
            aldex.leases.dp2.filt, 
            aldex.leases.dp3.filt)
```



## Bind datasets

```{r bind datasets}
# 18s
aldex_18s <- bind_rows(aldex.euk.filt, 
                       aldex.phyto.filt,
                       aldex.not.phyto.filt,
                       all_aldex.leases %>% filter(Kingdom %in% c("Eukaryotes", "Phytoplankton", "Other_protists"))) %>%
   mutate(Depth = str_sub(contrast, 1,3), .before = 3)

aldex_18s %>% 
  write_csv(paste0(folder_path,"data/feature_selection_18s/18s_aldex.csv"))

```


```{r clean R env}
rm(list = setdiff(ls(), c('folder_path','color_distance','color_depth','lease_control','pseq.euk.list.filt', "meta")))
source('~/R/phd/2019/script/my_functions.R')
source('~/R/phd/2019/script/theme_publication.R')
```

# Deseq2

### 18s 

Before running Deseq2 for 18S we remove samples with less than 500 sequences

log2FoldChange > 0 means it is more relatively abundant within the first factor and < 0 means it is more within the second factor

```{r}
deseq_fun_18s <- function(physeq, var1 ,value){
  # filter get data
  s <- physeq %>%
    get_tibble("sam_data") %>%
    filter(.[[var1]] == !!(value)) %>%
    pull(column_id)
  pseq.filt <- prune_samples(s, physeq)
  # keep taxa with prevalence > 1
  pseq <- filter_taxa(pseq.filt, function(x) sum(x > 0) > 1, TRUE)
  
  # Remove samples with less than 500 reads from phyloseq object
pseq <- prune_samples(sample_sums(pseq) >= 500, pseq)
  
# tax table
  TAXdf <-  pseq %>% get_tibble("tax_table", column_id = "ASV")
  #cut alpha level at 0.05. We might want to change this later to higher level of significance if we get heaps of otu responding
  #alpha can be changed to 0.005  or 0.01 if necessary
  alpha = 0.05
  # alpha = 0.001
  # alpha = 0.01
  lfc = 1 # lfcThreshold = 1, means 2-fold difference, (log2(2) = lfc). 
  # It tests for ASVs/OTUs that show significant effects of treatment on gene counts more than doubling or less than twice, because 2^1=2.
  #  lfc = 0.5849625 # means 1.5-fold difference, 
  
  #set up model 
  # sets it up so that the first factor will be on the bottom, so > 0 means it is higher in the second factor < 0 means higher in first factor
  sample_data(pseq)$site <- factor(sample_data(pseq)$site, levels = c("Control","lease_1","lease_2"))
  
  #look at lith Vs nonlith
  dds <- phyloseq_to_deseq2(pseq, ~site) # converts the phyloseq data into a DESeqDataSet with dispersions estimated
  dds <- DESeq2::DESeq(dds, test="Wald", fitType="parametric")
  # DESeq2::resultsNames(dds) # there are more comparisons 
  # 
  # # get the results
  # deseq_result <- DESeq2::results(dds,  lfcThreshold = lfc, alpha = alpha, altHypothesis = "greaterAbs") %>% 
  #   as.data.frame() %>% rownames_to_column('ASV')  %>%
  #   left_join(TAXdf %>% rownames_to_column('ASV'))
  
  #pull out model parameters for plotting
  # sigtab_filt = deseq_result %>% 
  #   filter(padj < alpha, abs(log2FoldChange) > 2) # logfoldchange = 2
  
  # for another comparisons 
  # sets factors up -> log2FoldChange > 0 means it is more relatively abundant within the first factor and < 0 means it is more within the second factor
  # So, a log2 fold change of 1.5 means that the gene’s expression is increased in the first factor by a multiplicative factor of 2^1.5 ≈ 2.82.
  deseq_result.control.lease1 <- DESeq2::results(dds,  
                                                lfcThreshold = lfc, 
                                                alpha = alpha, 
                                                altHypothesis = "greaterAbs", 
                                                contrast=c("site","lease_1", "Control")) %>% 
    as.data.frame() %>% 
    rownames_to_column('ASV')  %>%
    left_join(TAXdf) %>%
    mutate(contrast = "lease_1 vs control",
           Depth = glue::glue("{value}"), .before = 1) %>%
    tibble()
  
  deseq_result.control.lease2 <- DESeq2::results(dds,  
                                                lfcThreshold = lfc, 
                                                alpha = alpha, 
                                                altHypothesis = "greaterAbs", 
                                                contrast=c("site","lease_2","Control")) %>%
    as.data.frame() %>% 
    rownames_to_column('ASV')  %>%
    left_join(TAXdf) %>%
    mutate(contrast = "lease_2 vs control",
           Depth = glue::glue("{value}"), .before = 1) %>%
    tibble()

  
  deseq_result.leases <- DESeq2::results(dds,  
                                        lfcThreshold = lfc, 
                                        alpha = alpha, 
                                        altHypothesis = "greaterAbs", 
                                        contrast=c("site","lease_1","lease_2")) %>% 
    as.data.frame() %>% 
    rownames_to_column('ASV')  %>%
    left_join(TAXdf) %>%
    mutate(contrast = "lease_1 vs lease_2",
           Depth = glue::glue("{value}"), .before = 1) %>%
    tibble()
  
  #pull out model parameters for plotting
  deseq_results = bind_rows(deseq_result.control.lease1, 
                            deseq_result.control.lease2,
                            deseq_result.leases) 
  
  return(deseq_results)
}
```


```{r deseq2 - run}
pseq.euk.list.filt.18s <- pseq.euk.list.filt %>%
  keep(names(.) %in% c("Eukaryotes", "Phytoplankton", "Other_protists"))

# DP1
deseq.18s.dp1 <- pseq.euk.list.filt.18s %>% 
  map(., ~deseq_fun_18s(., "Depth", "DP1")) %>%
  bind_rows()

# DP2
deseq.18s.dp2 <- pseq.euk.list.filt %>% 
  map(., ~deseq_fun_18s(., "Depth", "DP2")) %>%
  bind_rows()

# DP3 # only possible for archaea and bacteria
deseq.18s.dp3 <- pseq.euk.list.filt %>% 
  map(., ~deseq_fun_18s(., "Depth", "DP3")) %>%
  bind_rows() 

```

# Bind datasets

```{r deseq2 - filtering}
# 18s
deseq_asv_sig.18s <- bind_rows(deseq.18s.dp1,
                       deseq.18s.dp2,
                       deseq.18s.dp3) %>% 
    filter(padj < 0.05 & abs(log2FoldChange) > 2) %>% # logfoldchange = 2
  rename(contrast_deseq = contrast) %>%
  mutate(method = "DESeq2",
         contrast1 = case_when(contrast_deseq == "lease_1 vs control" ~ "Control vs lease_1", 
                              contrast_deseq == "lease_2 vs control" ~ "Control vs lease_2",
                              TRUE ~ as.character(contrast_deseq)), .before = 2) %>%
  unite(contrast, Depth, contrast1, sep = " - ", remove = F) %>%
  select(-contrast1)

write_csv(deseq_asv_sig.18s, paste0(folder_path,"data/feature_selection_18s/18s_deseq.csv"))

# deseq_asv_sig = sigtab_filt %>%
#     arrange(desc(log2FoldChange)) %>%
#     mutate(Class = factor(Class,unique(Class))) # order by taxa for plotting

```

```{r clean R env}
rm(list = setdiff(ls(), c('folder_path','color_distance','color_depth','lease_control','pseq.euk.list.filt', "meta")))
source('~/R/phd/2019/script/my_functions.R')
source('~/R/phd/2019/script/theme_publication.R')
```


## loading tables
```{r data}
# 18s
aldex_18s <-
  read_csv(paste0(folder_path,"data/feature_selection_18s/18s_aldex.csv"))

deseq_18s <-
  read_csv(paste0(folder_path,"data/feature_selection_18s/18s_deseq.csv"))

```

## summary
```{r}
# total ASVs and funcitons
# Phytoplankton
phyto_18s <- aldex_18s %>% 
  filter(Kingdom == "Phytoplankton") %>%
  filter(effect < -1.5 | effect > 1.5) %>% 
  group_by(contrast) %>% 
  summarise(n()) %>%
  bind_rows(summarise(.,
                      across(where(is.numeric), sum),
                      across(where(is.character), ~"TOTAL"))) %>%
  add_row(contrast = 'Aldex2', .before = 1) %>%
  add_row(contrast = 'Deseq2') %>%
  bind_rows(deseq_18s %>% 
               filter(Kingdom == "Phytoplankton") %>%
              group_by(contrast) %>%
              summarise(n()) %>%
              bind_rows(summarise(.,
                        across(where(is.numeric), sum),
                        across(where(is.character), ~"TOTAL")))
            ) %>%
  rename(total = `n()`)

write_csv(phyto_18s,paste0(folder_path,"data/feature_selection_18s/phyto_18s.csv"))


# Other Protists
protists_18s <- aldex_18s %>% 
  filter(Kingdom == "Other_Protists") %>%
  filter(effect < -1.5 | effect > 1.5) %>% 
  group_by(contrast) %>% 
  summarise(n()) %>%
  bind_rows(summarise(.,
                      across(where(is.numeric), sum),
                      across(where(is.character), ~"TOTAL"))) %>%
  add_row(contrast = 'Aldex2', .before = 1) %>%
  add_row(contrast = 'Deseq2') %>%
  bind_rows(deseq_18s %>% 
               filter(Kingdom == "Other_Protists") %>%
              group_by(contrast) %>%
              summarise(n()) %>%
              bind_rows(summarise(.,
                        across(where(is.numeric), sum),
                        across(where(is.character), ~"TOTAL")))
            ) %>%
  rename(total = `n()`)

write_csv(protists_18s,paste0(folder_path,"data/feature_selection_18s/other_protists_18s.csv"))



```


```{r checking}
# tibble with ASVs common to both methods
deseq_18s %>% inner_join(aldex_18s, by = c("ASV", "contrast"))

# venn diagram
venn::venn(list(aldex = aldex_18s$ASV, deseq = deseq_18s$ASV), zcolor = "red, blue", ggplot = T)

deseq_18s %>% filter(if_any(everything(),~str_detect(.,"e_asv_1100")))
aldex_18s %>% filter(if_any(everything(),~str_detect(.,"e_asv_1100")))
```

# Ternary plots

Figure legend
FIG 6 Ternary plots of per-site mean relative abundances of OTUs. Panels represent the 12 most abundant phyla. Each point represents an OTU, and its position indicates the proportion of its relative abundance at the different sites along the urban gradient. Points closer to the ternary plot corners indicate that a greater proportion of the total relative abundance of this OTU was found in this particular environment. Point colors indicate the phylum of the OTU. The red lines inside the ternary plot indicate the 60% level of relative abundance of each of the sites. Only the 275 OTUs with a per-site mean relative abundance of more than 0.1% are shown. <https://journals.asm.org/doi/10.1128/mSystems.00087-17#>

### 18S

```{r function}
# https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggtern/

# https://online.ucpress.edu/elementa/article/9/1/00111/116754/Protist-communities-along-freshwater-marine

tern_plot <- function(pseq, Depth,title)
  {
  tax_df <- pseq %>% 
    get_tibble("tax_table", column_id = "ASV")

  name_taxa <- tax_df %>% unite("Taxa", Class, Family, sep = "_")
  n <- length(levels(as.factor(name_taxa$Taxa)))
  my_colors <- tibble(colors = friendly_cols[1:n],
                      name_taxa %>% distinct(Taxa))
  
  Kingdom <- tax_df %>% distinct(Kingdom) %>% pull
  
  asv_aldex <- aldex_18s %>% 
    filter(Depth == !!(Depth)) %>%
    filter(Kingdom == !!(Kingdom)) %>% 
    filter(effect < -1.5 | effect > 1.5) %>% 
    pull(ASV)
  
  asv_deseq <- deseq_18s %>% 
    filter(Depth == !!(Depth)) %>%
    filter(Kingdom == !!(Kingdom)) %>% 
    pull(ASV)

  tern_df <- pseq %>% 
    transform("compositional") %>%
    psmelt.dplyr() %>%
    rename(ASV = OTU)

  tern_df_dp <- tern_df %>%
    filter(Depth == !!(Depth)) %>% 
    group_by(ASV, site) %>% 
    summarise(value = mean(Abundance)) %>% 
    filter(ASV %in% c(asv_aldex, asv_deseq)) %>%
    pivot_wider(c(ASV), names_from = "site", values_from = "value") %>%
    mutate(sum = rowSums(across(where(is.numeric)))) %>%
    filter(sum > 0) %>%
    left_join(tax_df) %>%
    mutate(Taxa = str_c(Class, Family, sep = "_"),
           ASV = str_remove(ASV, "b_asv_")) %>%
    left_join(my_colors)
  
  col <- as.character(tern_df_dp$colors)
  names(col) <- as.character(tern_df_dp$Taxa)
  
  abund_tern <- tern_df_dp %>% 
    select(-Kingdom, -Taxa) %>%
    group_by(ASV, Phylum,Class,Order,Family,Genus, Species) %>%
    summarise(Control = Control/sum,
              lease_1 = lease_1/sum,
              lease_2 = lease_2/sum)
  
  # max <- tern_df_dp %>% 
  #   left_join(tern_df_dp  %>%
  #                select(ASV, lease_1, lease_2, Control) %>%
  #                pivot_longer(!ASV, names_to = 'max_site', values_to = 'value') %>%
  #                group_by(ASV) %>%
  #                 # If you're worried about keeping ties:
  #                slice(which.max(value))
  #              ) %>%
  #   dplyr::select(ASV, max_site, Phylum:Genus)
  
  # plot
  n  = length(tern_df_dp$ASV)   #Number of Data Points
  nv = 0.015  #Vertical Adjustment
  pn = position_nudge_tern(y=nv,x=-nv/2,z=-nv/2)
  
  plot <- tern_df_dp %>%
    ggtern(aes(x = lease_1, y = Control, z = lease_2)) + 
    geom_point(aes(size=(lease_1 + lease_2 + Control), color = Taxa)) +
    theme_Publication_3() +
    theme_showarrows() +
    theme_nomask() + #Allow Labels to Spool Over Edges
    geom_Lline(Lintercept=0.60,color="red") +
    geom_Tline(Tintercept=.6, colour='red') + 
    geom_Rline(Rintercept=.6, colour='red') +
    #  geom_text(position = pn, aes(label = ASV), check_overlap = T, hjust = 1, vjust = 1) +
    scale_size(name="Mean Relative Abundance") +
    scale_color_manual(name = "Taxonomic group", 
                       values = col, 
                       guide = guide_legend(label.theme = element_text(angle = 0, face = "italic"))) +
    labs(title = paste(title)) + 
    theme(plot.title = element_text(hjust=0.5))
  
  return(list(plot,abund_tern))
}

```


```{r plot ternary}
# prepare the data
load("pseq.euk.list.filt.RData") # filtered data
# plot ternary plots
surf_tern <- pseq.euk.list.filt %>% 
  map2(.,names(.), ~tern_plot(.,"DP1", paste0("Surface - ", .y)))
mid_tern <- pseq.euk.list.filt %>% 
  map2(.,names(.), ~tern_plot(.,"DP2", paste0("Intermediate - ", .y)))
bottom_tern <- pseq.euk.list.filt %>% 
  map2(.,names(.), ~tern_plot(.,"DP3", paste0("Bottom - ", .y)))

# get the a table with the ASV ASV in each site
surf_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_surface.csv")))

mid_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_middle.csv")))

bottom_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_bottom.csv")))
```

### Plot
```{r plot ternary}
tern_plots_fun <- function(df1, df2, df3, taxo){
  p1 <- df1 %>% pluck(taxo,1) 
  p2 <- df2 %>% pluck(taxo,1)
  p3 <- df3 %>% pluck(taxo,1) 

  # make a common legend
  df_colors <- bind_rows(p1$data %>% distinct(Taxa, colors),
  p2$data %>% distinct(Taxa, colors),
  p3$data %>% distinct(Taxa, colors)) %>%
    distinct(Taxa, colors)
  
  col <- as.character(df_colors$colors)
  names(col) <- as.character(df_colors$Taxa)
  
  # make a simple plot and get the legend
  plot <- df_colors %>% 
    mutate(n = 1) %>%  
    ggplot(aes(n, Taxa, color = Taxa)) +
    geom_point() +
    scale_color_manual(name = "Taxonomic group", values = col,
                       guide = guide_legend(label.theme = element_text(angle = 0, face = "italic"))) + theme_Publication_3()
  # get legend using ggpubr
  legend <- ggpubr::get_legend(plot)
  
  # combined plots
  a <- p1 + theme(legend.position = "none") 
  b <- p2 + theme(legend.position = "none")
  c <- p3 + theme(legend.position = "none")
  
  # plotmusing grid.arrange
  layout_matrix <- matrix(c(1, 2, 3, 4), nrow = 2, byrow = TRUE)
  plots <-  grid.arrange(a, b, c, legend, layout_matrix = layout_matrix)
  
  return(plots)
  
}

tern_plots_fun(surf_tern, mid_tern, bottom_tern, "Eukaryotes")
# save as pdf (18 x 10.2 in)

tern_plots_fun(surf_tern, mid_tern, bottom_tern, "Phytoplankton")
tern_plots_fun(surf_tern, mid_tern, bottom_tern, "Other_protists")
# save as pdf (15.32 x 9.8 in)


```


# Number of ASVs - bar plots

```{r}
surf_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_surface.csv")))

mid_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_middle.csv")))

bottom_tern %>% 
  map(.,2) %>% 
  map2(.,names(.), ~write_csv(.x,paste0(folder_path,"data/feature_selection_18s/tern_plot_",.y,"_bottom.csv")))
```

## Phytoplankton
```{r}
# phyto
phyto_surf <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_surface.csv"))
phyto_mid <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_middle.csv"))
phyto_bottom <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_bottom.csv"))

# choose one
site <- "lease_1"
site <- "lease_2"
site <- "Control"

# make table
n_asv_16s <- phyto_surf %>%
  arrange(desc(.data[[site]])) %>%
  filter(.data[[site]] >= 0.6) %>%
  mutate(depth = "surface") %>%
  bind_rows(phyto_mid %>%
              arrange(desc(.data[[site]])) %>%
              filter(.data[[site]] >= 0.6) %>%
              mutate(depth = "middle")) %>%
  bind_rows(phyto_bottom %>%
              arrange(desc(.data[[site]])) %>%
              filter(.data[[site]] >= 0.6) %>%
              mutate(depth = "bottom")) %>%
  group_by(depth, Phylum, Class,Order,Family, Genus) %>%
  count(Species,  sort = TRUE) 

n_asv_16s %>% write_csv(paste0(folder_path,"data/feature_selection_18s/n_asv_phyto_",site,".csv"))

```

## Other Protists
```{r}
# other prostis
other_surf <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_surface.csv"))
other_mid <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_middle.csv"))
other_bottom <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_bottom.csv"))

# choose one
site <- "lease_1"
site <- "lease_2"
site <- "Control"

# make table
n_asv_16s <- other_surf %>%
  arrange(desc(.data[[site]])) %>%
  filter(.data[[site]] >= 0.6) %>%
  mutate(depth = "surface") %>%
  bind_rows(other_mid %>%
              arrange(desc(.data[[site]])) %>%
              filter(.data[[site]] >= 0.6) %>%
              mutate(depth = "middle")) %>%
  bind_rows(other_bottom %>%
              arrange(desc(.data[[site]])) %>%
              filter(.data[[site]] >= 0.6) %>%
              mutate(depth = "bottom")) %>%
  group_by(depth, Phylum, Class,Order,Family, Genus) %>%
  count(Species,  sort = TRUE) 

n_asv_16s %>% write_csv(paste0(folder_path,"data/feature_selection_18s/n_asv_other_",site,".csv"))

```

#### 18S
```{r}
asv_surface <- surf_tern %>% 
  map(., 2) %>% 
  map2(.,names(.), ~mutate(., group = .y, depth = "surface")) %>%
  bind_rows()

asv_mid <- mid_tern %>% 
  map(., 2) %>% 
  map2(.,names(.), ~mutate(., group = .y, depth = "middle")) %>%
  bind_rows()

asv_bottom <- bottom_tern %>% 
  map(., 2) %>% 
  map2(.,names(.), ~mutate(., group = .y, depth = "bottom")) %>%
  bind_rows()

all_asvs <- asv_surface %>% 
  bind_rows(asv_mid, asv_bottom) 


all_asvs %>% 
  filter(group == "Eukaryotes" & lease_1 >= 0.6) %>%
  group_by(depth, Phylum, Class,Order,Family) %>%
  count(Genus,  sort = TRUE) 
  

plot_bar_dif_abund <- function(group, site){
  plot <- all_asvs %>% 
    arrange(desc(.data[[site]])) %>% 
    filter(.data[[site]] >= 0.6 & group == !!(group)) %>% 
    # filter(if_any(everything(), ~str_detect(.,"unknown")))
   # select(-ASV) %>%
    group_by(depth, Phylum, Class,Order,Family, Genus) %>%
    count(Species,  sort = TRUE) %>%
    ungroup() %>%
     mutate(Genus =  str_c(Class, Genus, Species, sep = "_"),
            depth = factor(depth, levels = c("surface", "middle", "bottom")),
            Genus = tidytext::reorder_within(Genus, n, depth)) %>%
   # mutate(Genus = fct_reorder(Genus,desc(n))) %>%
    ggplot(aes(x = Genus, y = n, fill = depth)) +
    geom_col(alpha = 0.7) +
     #      geom_text(aes(label = str_c(Order, Family, sep = "_")), y=0.1, hjust=0) +
           labs(y = "number of ASVs",
                x = "Taxonomic group") +
          facet_grid(depth~., scales = "free") +
          coord_flip() +
          tidytext::scale_x_reordered() +
          labs(title = paste(group, " - ", site)) +
           scale_fill_manual(values = c("surface" = "darkorange", "middle" = "purple", "bottom" = "cyan4")) +
          theme(legend.position = "bottom",
                axis.text.y = element_text(face = "italic"))
  return(plot)
}

p1 <- plot_bar_dif_abund("Eukaryotes", "lease_1")
p2 <- plot_bar_dif_abund("Eukaryotes", "lease_2")
p3 <- plot_bar_dif_abund("Eukaryotes", "Control")

all_terns <- p1 / p2 / p3 + patchwork::plot_layout(guides='collect') & theme(legend.position = "bottom")

ggsave(all_terns, 
       filename = paste0(folder_path,"plots/feature_selection_18s/n_asvs_eukaryotes.tiff"),
       compression = "lzw", height = 16, width = 12)

p1 <- plot_bar_dif_abund("Phytoplankton", "lease_1")
p2 <- plot_bar_dif_abund("Phytoplankton", "lease_2")
p3 <- plot_bar_dif_abund("Phytoplankton", "Control")

all_terns <- p1 / p2 / p3 + patchwork::plot_layout(guides='collect') & theme(legend.position = "bottom")

ggsave(all_terns, 
       filename = paste0(folder_path,"plots/feature_selection_18s/n_asvs_phyto.tiff"),
       compression = "lzw", height = 10, width = 7.5)

p1 <- plot_bar_dif_abund("Other_protists", "lease_1")
p2 <- plot_bar_dif_abund("Other_protists", "lease_2")
p3 <- plot_bar_dif_abund("Other_protists", "Control")


all_terns <- p1 / p2 / p3 + patchwork::plot_layout(guides='collect') & theme(legend.position = "bottom")

ggsave(all_terns, 
       filename = paste0(folder_path,"plots/feature_selection_18s/n_asvs_other.tiff"),
       compression = "lzw", height = 10.5, width = 9.6)

```


