---
title: "ML - Regression"
author: "Ricardo Silva"
date: "02/09/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      eval=FALSE, 
                      message=FALSE, 
                      warning=FALSE)
```

```{r CLEAR EVERYTHING, eval=FALSE, include=FALSE}
# unload all non-base packages
lapply(names(sessionInfo()$otherPkgs), function(pkgs)
  detach(
    paste0('package:', pkgs),
    character.only = T,
    unload = T,
    force = T
))

rm(list=ls())
```

# Packages and initial steps

```{r load libraries, message=FALSE, warning=FALSE, include=FALSE}
# load the packages
# main packs to start with
pcks <- c('phyloseq', 'microbiome','microbiomeutilities', "patchwork",'tidymodels',"themis",'ggtern','tidyverse')

if(sum(as.numeric(!pcks %in% installed.packages())) != 0){
  installation <- pcks[!pcks %in% installed.packages()]
  for(i in 1:length(installation)) {
    install.packages(installation, dependencies = T)
    break()}
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
} else {
  suppressPackageStartupMessages(
  sapply(pcks,require,character.only = T)
) 
}

rm(pcks)
```

```{r packages and functions}
# set working directory and seed
setwd("C:/Users/rrocha/Documents/R/phd/2019/data")
folder_path <- "C:/Users/rrocha/Documents/R/phd/2019/output/"

source('~/R/phd/2019/script/my_functions.R')
source('~/R/phd/2019/script/theme_publication.R')
theme_set(theme_Publication_3())

# set colors
color_distance <- c("D0" = "deepskyblue2", 'D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
# color_distance1 <- c('D1' = 'brown1',  "D2" = "forestgreen", "D3" = "#440154FF", "D4" = "burlywood")
color_depth <- viridis::viridis(n=5)
lease_control <- friendly_cols[1:3]
#scales::show_col(viridis::viridis(n=3))

# check the working directory
current_directory <- getwd()
paste("Working in directory:", current_directory)

# create dir
# dir.create(paste0(folder_path,"plots/random_forest_18s"))
# dir.create(paste0(folder_path,"data/random_forest_18s"))
```

```{r load data}
# load object in RData format
# 16S and 18S
# load("pseq.euk.list.RData")
# load("pseq.euk.list.filt.RData") # filtered data
load("pseq.euk.list.clr.RData") # clr transformed data
pseq.euk.list.clr

```

# 2 - Random forest

## function
```{r set model}
# Specify the model and engine
rf_model <- 
  rand_forest(mtry = tune(), 
              min_n = tune(), 
              trees = 501) %>% 
  set_mode("classification")  %>%
  set_engine("randomForest", 
             importance = TRUE) 

```

```{r big function}
# big function of model fitting and predicting
predict_rf <- function(pseq, variable) {
  
  set.seed(4848)
  
  # get ASV table
  clr_otu <- pseq %>% 
  get_tibble("otu_table", column_id = "ASV") %>% 
  column_to_rownames("ASV") %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column("sample_id") 

  # meta table
  meta.table <- pseq %>% 
    get_tibble("sam_data", column_id = "sample_id") %>%
    mutate_if(is.character, as.factor)
  
  # tax table
  tax.df <- pseq %>% 
    get_tibble("tax_table", column_id = "ASV")
  
  # Kingdom name = plot title
  kingdom <- tax.df %>% 
    pull(Kingdom) %>% 
    unique()
  
  # prepare the data
# join datasets
df <- clr_otu %>% 
  left_join(meta.table %>% 
              select(sample_id, all_of(variable))) %>%
  select(-sample_id) %>%
  rename(variable = variable)

df_split <- df %>% initial_split(prop = 3/4, strata = variable)

# extract training and testing sets
df_train <- df_split %>% training
df_test <- df_split %>% testing

# create CV object from training data
df_cv <- df_train %>% vfold_cv(v = 10, repeats = 3, strata = variable)


rf_rec <- 
  recipe(formula = variable ~ ., data = df_train) %>%
  # normalize the data (center and scale all predictors = mean to zero and standard deviation of one)
  step_normalize(all_predictors()) %>%
  # Synthetic Minority Over-sampling Technique
  step_smote(variable)

rf_workflow <- workflow() %>%
  # add the model
  add_model(rf_model) %>%
  # add recipe
  add_recipe(rf_rec)

# choose the metrics for the model
members_metrics <- metric_set(roc_auc, accuracy, kap, sensitivity, specificity)

# training and tune
rf_tune_results_rf <- rf_workflow %>%
  tune_grid(resamples = df_cv, #CV object
            grid = 20, # grid of values to try
            metrics = members_metrics,
            control = control_grid(save_pred = T)
  )


# Finalize workflow
collect_metrics(rf_tune_results_rf, summarize = TRUE)

# show_best(rf_tune_results_rf, "accuracy")

final_rf <- rf_workflow %>%
  finalize_workflow(select_best(rf_tune_results_rf, "roc_auc"))

# Evaluate the model on the test set
# fit
# rf_rs <-
#   fit_resamples(final_rf,
#     resamples = df_cv,
#     metrics = members_metrics,
#     control = control_resamples(save_pred = TRUE)
#   )
# 
# collect_metrics(rf_rs)
# 
# rf_rs %>%
#   conf_mat_resampled()

# last fit
# will train the model using training data, and produce evaluations based on the test set
rf_last_fit <- final_rf %>%
  # fit on the training set and evaluate on test set
  last_fit(df_split,
           metrics = members_metrics)

confucion_matrix <- collect_predictions(rf_last_fit) %>%
                        conf_mat(variable, .pred_class)

# confucion_matrix %>% autoplot(type = "heatmap")

# evaluating
rf_performance <- rf_last_fit %>% collect_metrics()
  
  # Variable importance
  # ASVs vip plot
  plot.vip <- rf_last_fit %>%
    extract_fit_parsnip() %>%
    vip::vip(geom =  "point", 
             aesthetics = list(color = "midnightblue", size = 3)) +
    ggtitle(paste(kingdom))
  
  # vip ASVs table
  table.vip <- rf_last_fit %>%
    extract_fit_parsnip() %>%
    vip::vi() %>%
    rename(ASV = Variable) %>%
    left_join(tax.df)

return(list(model = rf_last_fit, 
            conf.matrix = confucion_matrix , 
            performance = rf_performance, 
            vip.plot = plot.vip, 
            vip.table = table.vip))

}
```

## data preparation - 18s
```{r}
# # ASVs diff abundant
# aldex_18s <-
#   read_csv(paste0(folder_path,"data/feature_selection_18s/18s_aldex.csv")) %>%
#   filter(effect < -1.5 | effect > 1.5)
# 
# deseq_18s <-
#   read_csv(paste0(folder_path,"data/feature_selection_18s/18s_deseq.csv"))
# 
# venn::venn(list(aldex_18s$ASV, deseq_18s$ASV))
# 
# ASVs_select <- aldex_18s %>%
#   full_join(deseq_18s, by="ASV") %>%
# #  filter(str_detect(ASV, "e_")) %>%
#   mutate(ASV = sort(ASV)) %>%
#   distinct(ASV) %>%
#   pull
```

## run models
```{r}
# variables
# vars <- c("NOX", "NH4", "oxygen", "turbidity")

# doParallel
cores <- parallel::detectCores(logical = FALSE)
cl <- parallel::makePSOCKcluster(cores)
doParallel::registerDoParallel(cores = cl)

# run the function for each variable
set.seed(4848)
rf_all <- predict_rf(pseq.euk.list.clr[[1]], "site") 
# rf_all[[2]]$.workflow
rf_phyto <- predict_rf(pseq.euk.list.clr[[2]], "site") 
# rf_phyto[[2]]$.workflow
rf_other <- predict_rf(pseq.euk.list.clr[[3]], "site") 
# rf_other[[2]]$.workflow

save(rf_phyto, file = paste0(folder_path,'data/random_forest_18s/rf_phyto.RData'))
save(rf_other, file = paste0(folder_path,'data/random_forest_18s/rf_other.RData'))

# get the model specification
load(paste0(folder_path,'data/random_forest_18s/rf_phyto.RData'))
rf_phyto$model %>% extract_fit_parsnip() %>% pluck("spec")

load(paste0(folder_path,'data/random_forest_18s/rf_other.RData'))
rf_other$model %>% extract_fit_parsnip() %>% pluck("spec")

# # imbalanced learning - no down/up-sampling
# set.seed(4848)
# rf_all_imbalanced <- predict_rf(pseq.euk.list.clr[[1]], "site") 
# # rf_all_imbalanced[[1]]$.workflow
# rf_phyto_imbalanced <- predict_rf(pseq.euk.list.clr[[2]], "site") 
# # rf_phyto_imbalanced[[1]]$.workflow
# rf_other_imbalanced <- predict_rf(pseq.euk.list.clr[[3]], "site") 
# # rf_other_imbalanced[[1]]$.workflow

```

## get results
```{r}
# model performances
perf_rfs <- rf_all %>% 
  pluck("performance") %>%
  add_row(.metric = "Eukaryotes", .before = 1) %>%
  bind_rows(rf_phyto %>% 
              pluck("performance") %>%
              add_row(.metric = "Phytoplankton", .before = 1),
            rf_other %>% 
              pluck("performance") %>%
              add_row(.metric = "Other Protists", .before = 1))

perf_rfs %>% write_csv(paste0(folder_path,'data/random_forest_18s/performance_models.csv'))

# vip plots
p_all <- rf_all %>% 
  pluck("vip.plot") 
p_all %>%  ggsave(filename = paste0(folder_path, "plots/random_forest_18s/vip_plot_all.tiff"), 
                  compression = "lzw",  width = 8.55 , height =  5.58)

p_phyt <- rf_phyto %>% 
  pluck("vip.plot") 
p_phyt %>%  ggsave(filename = paste0(folder_path, "plots/random_forest_18s/vip_plot_phyto.tiff"), 
                  compression = "lzw",  width = 8.55 , height =  5.58)

p_other <- rf_other %>% 
  pluck("vip.plot") 
p_other %>%  ggsave(filename = paste0(folder_path, "plots/random_forest_18s/vip_plot_other.tiff"), 
                  compression = "lzw",  width = 8.55 , height =  5.58)

vip.plots <- p_all + p_phyt + p_other
# 14.9 x 5.17 in
vip.plots %>% ggsave(filename = paste0(folder_path, "plots/random_forest_18s/vip_plots.tiff"), compression = "lzw")

# save table with vip ASV
asv.vip.tb <- rf_all %>% 
  pluck("vip.table") %>%
  bind_rows(rf_phyto %>% 
              pluck("vip.table"),
            rf_other %>% 
              pluck("vip.table"))

asv.vip.tb %>% write_csv(paste0(folder_path,'data/random_forest_18s/asv_table.csv'))


# top 40
asv.vip.tb %>%
  group_by(Kingdom) %>%
  slice_max(Importance, n = 40) %>%
  write_csv(paste0(folder_path,'data/random_forest/vip_table_top40.csv'))

```

results = Glasl et al, 2019

## Bar plot

#### 18s
```{r}
vip_18s <- read_csv(paste0(folder_path,'data/random_forest_18s/asv_table.csv'))

imp_plot <- function(df){
imp_plot <- df %>%
  mutate(name = fct_reorder(ASV, Importance),
         label = str_c(Class, Order, Family, Genus, Species, sep = "_"),
         label = str_remove(label, c("_unknown", "unknown_"))) %>%
  slice_head(n=10) %>%
  ggplot(aes(x=name, y=Importance)) +
    geom_bar(stat="identity", fill="darkseagreen1", alpha=.8, width=.4) +
    geom_text(aes(label = label), y=0.1, hjust=0, fontface = "italic") +
    coord_flip() +
    xlab("")
  return(imp_plot)
}

# nest_vip_line <- rf_NH4[[4]] %>%
#   mutate(env_par = "NH4") %>%
#   bind_rows(rf_NOX[[4]] %>% 
#   mutate(env_par = "NOX"),
#   rf_oxygen[[4]] %>% 
#   mutate(env_par = "oxygen"),
#   rf_turb[[4]] %>% 
#   mutate(env_par = "turbidity")) %>%
#   group_by(env_par) %>%
#   nest %>%
#   mutate(plots = map2(data, env_par, ~imp_plot(.x) + 
#                         ggtitle(paste(.y))
#                       ))


# vip_18s <- wrap_plots(nest_vip_line$plots) + plot_annotation(title = "Bacterial Community")
# ggsave(vip_18s, filename = paste0(folder_path,"plots/random_forest/vip_rf_16s.tiff"), compression = "lzw", width = 21, height = 12)
# 
# walk2(nest_vip_line$env_par, nest_vip_line$plots,
#     ~ggsave(filename = paste0(folder_path,"plots/random_forest/vip_rf_",.x, ".tiff"), 
#             plot = .y, 
#             compression = "lzw"))


vip.plot <- vip_18s %>% 
  group_by(Kingdom) %>%
  nest %>%
  mutate(plots = map2(data, Kingdom, ~imp_plot(.x) + 
                        ggtitle(paste(.y))
                      ))

# 10.6 x 6.07 in
vip.plot$plots %>% map2(.,c('Eukaryota','Phytoplankton','Other_Protists') ,~ggsave(., filename = paste0(folder_path,"plots/random_forest_18s/vip_rf_top_10_",.y,".tiff"), 
                                  compression = "lzw"))

# all together
vip_picrust <- wrap_plots(vip.plot$plots, nrow = 3) + plot_annotation(title = "Eukaryotic community")
ggsave(vip_picrust, filename = paste0(folder_path,"plots/random_forest_18s/vip_rf_top10.tiff"), compression = "lzw", width = 9.5, height = 11)


```

## Ternary

### 18s
```{r}
# https://yutannihilation.github.io/allYourFigureAreBelongToUs/ggtern/
# https://online.ucpress.edu/elementa/article/9/1/00111/116754/Protist-communities-along-freshwater-marine

vip_18s <- read_csv(paste0(folder_path,'data/random_forest_18s/asv_table.csv'))

tern_plot_rf <- function(pseq, n_importance)
  {

  tax_df <- pseq %>% 
    get_tibble("tax_table", column_id = "ASV")
  
   taxa <- tax_df %>% 
    pull(Kingdom) %>% 
    unique()
   
  name_taxa <- tax_df %>% unite("Taxa", Class,Order,Species, sep = "_")
  n <- length(levels(as.factor(name_taxa$Taxa)))
  my_colors <- tibble(colors = friendly_cols[1:n],
                      name_taxa %>% distinct(Taxa))
  
  tern_df <- pseq %>% 
    transform("compositional") %>%
    psmelt.dplyr() %>%
    rename(ASV = OTU)
 
  asv_rf <- vip_18s %>% filter(Kingdom == !!(taxa)) %>% slice_max(Importance, n = n_importance) %>% distinct(ASV) %>% pull
  
 df_to_join <- vip_18s %>% 
                filter(Kingdom == !!(taxa)) %>%
                mutate(ASV = str_remove(ASV, "e_asv_")) %>%
                select(ASV, Importance)

  tern_df_dp <- tern_df %>%
 #   filter(Depth == !!(Depth)) %>% 
    group_by(ASV, site) %>% 
    summarise(value = mean(Abundance)) %>% 
    filter(ASV %in% asv_rf) %>%
    pivot_wider(c(ASV), names_from = "site", values_from = "value") %>%
    mutate(sum = rowSums(across(where(is.numeric)))) %>%
    filter(sum > 0) %>%
    left_join(tax_df) %>%
    mutate(Taxa = str_c(Class,Order,Species, sep = "_"),
           ASV = str_remove(ASV, "e_asv_")) %>%
    left_join(my_colors)
  
  col <- as.character(tern_df_dp$colors)
  names(col) <- as.character(tern_df_dp$Taxa)
  
  abund_tern <- tern_df_dp %>%
    select(-Taxa) %>%
    group_by(ASV, Kingdom, Phylum,Class,Order,Family,Genus, Species) %>%
    summarise(Control = Control/sum,
              lease_1 = lease_1/sum,
              lease_2 = lease_2/sum) %>%
    left_join(df_to_join)  %>%
    arrange(desc(Importance))
  
  # max <- tern_df_dp %>% 
  #   left_join(tern_df_dp  %>%
  #                select(ASV, lease_1, lease_2, Control) %>%
  #                pivot_longer(!ASV, names_to = 'max_site', values_to = 'value') %>%
  #                group_by(ASV) %>%
  #                 # If you're worried about keeping ties:
  #                slice(which.max(value))
  #              ) %>%
  #   dplyr::select(ASV, max_site, Phylum:Genus)
  
  # plot
  n  = length(tern_df_dp$ASV)   #Number of Data Points
  nv = 0.015  #Vertical Adjustment
  pn = position_nudge_tern(y=nv,x=-nv/2,z=-nv/2)
  
  plot <- tern_df_dp %>%
    ggtern(aes(x = lease_1, y = Control, z = lease_2)) + 
    geom_point(aes(size=(lease_1 + lease_2 + Control), color = Taxa)) +
    theme_Publication_3() +
    theme_showarrows() +
    theme_nomask() + #Allow Labels to Spool Over Edges
    geom_Lline(Lintercept=0.60,color="red") +
    geom_Tline(Tintercept=.6, colour='red') + 
    geom_Rline(Rintercept=.6, colour='red') +
    #  geom_text(position = pn, aes(label = ASV), check_overlap = T, hjust = 1, vjust = 1) +
    scale_size(name="Mean Relative Abundance") +
    scale_color_manual(name = "Taxonomic group", 
                       values = col, 
                       guide = guide_legend(label.theme = element_text(angle = 0, face = "italic"))) +
    labs(title = paste(taxa)) + 
    theme(plot.title = element_text(hjust=0.5))
  
  return(list(plot, abund_tern))
}
```


```{r}
rm(tax_df)
# prepare the data
load("pseq.euk.list.filt.RData") # filtered data
# plot ternary plots
pseq.euk.list.filt %>% map(., ~tern_plot_rf(.,n_importance = 10))

vip_all <- tern_plot_rf(pseq.euk.list.filt[[1]], 
                             n_importance = 10) 
vip_phy <- tern_plot_rf(pseq.euk.list.filt[[2]], 
                             n_importance = 20) 
vip_other <- tern_plot_rf(pseq.euk.list.filt[[3]], 
                             n_importance = 20)  

# make a dataframe
df_vip_18s <- bind_rows(vip_all[[2]], vip_phy[[2]], vip_other[[2]])
df_vip_18s %>% write_csv(paste0(folder_path,"data/random_forest_18s/tern_vip_rf.csv"))


```

### site vs depth - top 20 VIP
```{r}
# site
vip_phy <- tern_plot_rf(pseq.euk.list.filt[[2]], 
                             n_importance = 20) 
vip_other <- tern_plot_rf(pseq.euk.list.filt[[3]], 
                             n_importance = 20) 

# depth
vip_phy_depth <- tern_plot_rf_depth(pseq.euk.list.filt[[2]], 
                             n_importance = 20) 
vip_other_depth <- tern_plot_rf_depth(pseq.euk.list.filt[[3]], 
                             n_importance = 20) 

# function to plot
tern_plots_dist_depth <- function(df1, df2){
  p1 <- df1 %>% pluck(1) 
  p2 <- df2 %>% pluck(1)
  
  # make a common legend
  df_colors <- bind_rows(p1$data %>% distinct(Taxa, colors),
  p2$data %>% distinct(Taxa, colors)) %>%
    distinct(Taxa, colors)
  
  col <- as.character(df_colors$colors)
  names(col) <- as.character(df_colors$Taxa)
  
  # make a simple plot and get the legend
  plot <- df_colors %>% 
    mutate(n = 1) %>%  
    ggplot(aes(n, Taxa, color = Taxa)) +
    geom_point() +
    scale_color_manual(name = "Taxonomic group", values = col,
                       guide = guide_legend(ncol = 2,  label.theme = element_text(angle = 0, face = "italic"))) + theme_Publication_3()
  # get legend using ggpubr
  legend <- ggpubr::get_legend(plot)
  
  # combined plots
  a <- p1 + theme(legend.position = "none") 
  b <- p2 + theme(legend.position = "none")

  # plotmusing grid.arrange
  layout_matrix <- matrix(c(1,2,3,3), nrow = 2, byrow = TRUE)
  plots <-  grid.arrange(a, b, legend, layout_matrix = layout_matrix)
  
  return(plots)
  
}

phy_tern_dist_depth <- tern_plots_dist_depth(vip_phy,vip_phy_depth)
other_tern_dist_depth <- tern_plots_dist_depth(vip_other,vip_other_depth)
# save as pdf (15 x 10.2 in)

```

# Venn Diagram
## 18s
```{r}
library(ggVennDiagram)

rf_venn <- read_csv(paste0(folder_path,'data/random_forest_18s/asv_table.csv')) 
aldex_venn <- read_csv(paste0(folder_path,'data/feature_selection_18s/18s_aldex.csv')) %>%
  filter(effect < -1.5 | effect > 1.5)
deseq_venn <- read_csv(paste0(folder_path,'data/feature_selection_18s/18s_deseq.csv'))

venn_plot <- function(taxa, n){
  asvs <- list(RF = rf_venn %>% 
              filter(Kingdom == taxa) %>% 
              slice_max(Importance, n = n) %>% 
              pull(ASV),
       ALDEx2 = aldex_venn %>% 
              filter(Kingdom == taxa) %>% 
              pull(ASV),
       DESeq2 = deseq_venn %>% 
              filter(Kingdom == taxa) %>% 
              pull(ASV))
  
  v.plot <- ggVennDiagram(asvs) + scale_fill_gradient(low="white",high = "red") +
    ggtitle(paste0(taxa," - ",n," ASVs"))
  intersections <- process_region_data(Venn(asvs))
  
  return(list(plot = v.plot, int.table = intersections))
}

venn_all <- venn_plot("Eukaryota", n = 10)
venn_all_40 <- venn_plot("Eukaryota", n = 40)

venn_phy <- venn_plot("Phytoplankton", n = 10)
venn_phy_40 <- venn_plot("Phytoplankton", n = 40)

venn_other <- venn_plot("Other_Protists", n = 10)
venn_other_40 <- venn_plot("Other_Protists", n = 40)


# save as tiff
# Phytoplankton
p1 <- venn_phy %>% 
  pluck("plot") %>% 
  ggsave(filename = paste0(folder_path,'plots/feature_selection_18s/venn_common_asvs_top_10_phyto.tiff'), compression = "lzw")

p2 <- venn_phy_40 %>% 
  pluck("plot") %>% 
  ggsave(filename = paste0(folder_path,'plots/feature_selection_18s/venn_common_asvs_top_40_phy.tiff'), compression = "lzw")

# other Protists
p3 <- venn_other %>% 
  pluck("plot") %>% 
  ggsave(filename = paste0(folder_path,'plots/feature_selection_18s/venn_common_asvs_top_10_other.tiff'), compression = "lzw")

p4 <- venn_other_40 %>% 
  pluck("plot") %>% 
  ggsave(filename = paste0(folder_path,'plots/feature_selection_18s/venn_common_asvs_top_40_other.tiff'), compression = "lzw")

p_all <- (p1 + p2) / (p3 + p4) & theme(legend.position = "none")

ggsave(p_all, filename = paste0(folder_path,'plots/feature_selection_18s/all_venn_common_asvs.tiff'), compression = "lzw", height = 10 , width = 15)

```

### common ASVS

Get ASVs biomarkers in at least 2 methods
```{r}
# common ASVs in at least two methods
load("pseq.euk.list.clr.RData")

# phytoplankton
tax_phyto_df <- pseq.euk.list.clr[[2]] %>% get_tibble("tax_table", column_id = "ASV")
phyto_asv <- venn_phy_40[[2]] %>% 
   unnest(cols = item) %>%
   select(-component, -id, -count) %>% 
   rename(ASV = item) %>%
   left_join(tax_phyto_df)

write_csv(phyto_asv, file = paste0(folder_path,'data/random_forest_18s/common_asv_table_phytoplankton.csv'))

ASV_common_phyto <- phyto_asv %>%
   group_split(name) %>%
   set_names(., ~ letters[seq_along(.)]) %>%
   keep(names(.) %in% c("b","e", "f", "g")) %>%
   map_df(., bind_rows) %>%
  pull(ASV)

phyto_asv %>%
   group_split(name) %>%
   set_names(., ~ letters[seq_along(.)]) %>%
   keep(names(.) %in% c("b","e", "f", "g")) %>%
   map_df(., bind_rows) %>%
  write_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_phytoplankton_3methods.csv'))

# Other protists
tax_other_df <- pseq.euk.list.clr[[3]] %>% get_tibble("tax_table", column_id = "ASV")
other_asv <- venn_other_40[[2]] %>% 
   unnest(cols = item) %>%
   select(-component, -id, -count) %>% 
   rename(ASV = item) %>%
   left_join(tax_other_df) 


write_csv(other_asv, file = paste0(folder_path,'data/random_forest_18s/common_asv_table_other_protists.csv'))

ASV_common_other <- other_asv %>%
   group_split(name) %>%
   set_names(., ~ letters[seq_along(.)]) %>%
   keep(names(.) %in% c("b","e", "f", "g")) %>%
   map_df(., bind_rows) %>%
   pull(ASV)

other_asv %>%
   group_split(name) %>%
   set_names(., ~ letters[seq_along(.)]) %>%
   keep(names(.) %in% c("b","e", "f", "g")) %>%
   map_df(., bind_rows) %>%
  write_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_other_protists_3methods.csv'))


```

```{r bind ternary table}
# bind with ternary plot table
# Phytoplankton
common_phyto <- read_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_phytoplankton_3methods.csv'))

tern_phyto_surc <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_surface.csv")) %>%
  mutate(Depth = "surface")
tern_phyto_middle <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_middle.csv")) %>%
  mutate(Depth = "middle")
tern_phyto_bottom <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Phytoplankton_bottom.csv")) %>%
  mutate(Depth = "bottom")

all_phyto_tern <- tern_phyto_surc %>% bind_rows(tern_phyto_middle, tern_phyto_bottom)

common_phyto %>% 
  left_join(all_phyto_tern) %>% 
  write_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_phytoplankton_3methods_abundance.csv'))

# Other protists
common_other <- read_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_other_protists_3methods.csv'))

tern_other_surc <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_surface.csv")) %>%
  mutate(Depth = "surface")
tern_other_middle <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_middle.csv")) %>%
  mutate(Depth = "middle")
tern_other_bottom <- read_csv(paste0(folder_path,"data/feature_selection_18s/tern_plot_Other_protists_bottom.csv")) %>%
  mutate(Depth = "bottom")

all_other_tern <- tern_other_surc %>% bind_rows(tern_other_middle, tern_other_bottom)

common_other %>% 
  left_join(all_other_tern) %>% 
  write_csv(file = paste0(folder_path,'data/random_forest_18s/common_asv_table_other_protists_3methods_abundance.csv'))
```


## Distance profile - top 5 VIP

```{r}
# function to plot the top 9 VIPs ASVs
asvs_distance <- function(pseq, asvs_common, taxo){

tax_df <- pseq %>% 
  get_tibble("tax_table", column_id = "ASV")

# lease 1
pseq1 <- pseq %>% 
  subset_samples(site %in% c('lease_1', 'Control'))
otu_df1 <- pseq1 %>% 
  get_tibble("otu_table", column_id = "ASV")

otu_rf1 <- otu_df1 %>% 
  filter(ASV %in% asvs_common) %>%
  column_to_rownames("ASV") 
tax_rf1 <- tax_df %>% 
  filter(ASV %in% asvs_common)  %>%
  column_to_rownames("ASV")

phy_rf_lease1 <- phyloseq(otu_table(as.matrix(as.data.frame(otu_rf1)), taxa_are_rows = TRUE),
                     tax_table(as.matrix(tax_rf1)), sample_data(meta_lease1))

rf_df_1 <- phy_rf_lease1 %>% 
  psmelt.dplyr() %>%
  rename(ASV = OTU)


p_l1 <- rf_df_1  %>%
    mutate(ASV = factor(ASV, asvs_common),
      Distance_m = round(Distance_m, 1),
           Distance_m = as.factor(Distance_m)) %>%
  mutate(ASV = str_c(ASV, Phylum, Class, Order, Family, sep = "_"),
         ASV = str_remove(ASV, "e_asv_")) %>%
  ggplot(aes(x = Distance_m, y = Abundance, group = Depth)) + 
    geom_jitter(aes(color = Depth),
                alpha = 0.6,
                show.legend = T, 
                width = 0.1) + 
#    geom_line(aes(linetype = Depth, color = Depth)) +
    stat_smooth(aes(color = Depth),
                se = F, size = 1,
                show.legend = F, alpha = 0.3) +
    # scale_color_manual(values = asvvarpal, labels = asvvarlabels) + 
    # scale_x_continuous(breaks = cumnum,
    #                    name = 'Month',
    #                    labels = str_to_title(month.order)) +
    facet_wrap(~ASV, scales = "free_y") + 
    ylab('Centered log ratio') +
    scale_color_manual(labels = unique(meta_lease1$Depth1), values = c(color_depth)) +
    geom_vline(xintercept = '82.8', linetype="dashed", color = "red") +
    ggtitle(paste0('Lease 1: ', taxo)) +
  theme(legend.position = "bottom",
        strip.text = element_text(face = "italic", size = 8),
        axis.title.x =  element_blank())

# lease 2
pseq2 <- pseq %>% 
  subset_samples(site %in% c('lease_2', 'Control'))
otu_df2 <- pseq2 %>% 
  get_tibble("otu_table", column_id = "ASV")

otu_rf2 <- otu_df2 %>% 
  filter(ASV %in% asvs_common) %>% 
  column_to_rownames("ASV")
tax_rf2 <- tax_df %>% 
  filter(ASV %in% asvs_common)  %>%
  column_to_rownames("ASV")

phy_rf_lease2 <- phyloseq(otu_table(as.matrix(as.data.frame(otu_rf2)), taxa_are_rows = TRUE),
                     tax_table(as.matrix(tax_rf2)), sample_data(meta_lease2))

rf_df_2 <- phy_rf_lease2 %>% 
  psmelt.dplyr() %>%
  rename(ASV = OTU) 

p_l2 <- rf_df_2  %>%
    mutate(ASV = factor(ASV, asvs_common),
           Distance_m = round(Distance_m, 1),
           Distance_m = as.factor(Distance_m)) %>%
  mutate(ASV = str_c(ASV, Phylum, Class, Order, Family, sep = "_"),
         ASV = str_remove(ASV, "e_asv_")) %>%
  ggplot(aes(x = Distance_m, y = Abundance, group = Depth)) + 
    geom_jitter(aes(color = Depth),
                alpha = 0.6,
                show.legend = T, 
                width = 0.1) + 
#    geom_line(aes(linetype = Depth, color = Depth)) +
    stat_smooth(aes(color = Depth),
                se = F, size = 1,
                show.legend = F, alpha = 0.3) +
    # scale_color_manual(values = asvvarpal, labels = asvvarlabels) + 
    # scale_x_continuous(breaks = cumnum,
    #                    name = 'Month',
    #                    labels = str_to_title(month.order)) +
    # facet_wrap(~str_c(ASV,Class, Family, Genus, sep = '_'), scales = "free_y") + 
    facet_wrap(~ASV, scales = "free_y") +
    ylab('Centered log ratio') +
    scale_color_manual(labels = unique(meta_lease2$Depth1), values = c(color_depth)) +
    geom_vline(xintercept = '84.2', linetype="dashed", color = "red") +
    ggtitle(paste0('Lease 2: ', taxo)) +
  theme(legend.position = "bottom",
        strip.text = element_text(face = "italic", size = 8))

return(list(p_l1, p_l2))
}
```

### 18s
```{r}
# split meta data to be used in the function
meta_lease2 <- read_csv(paste0(getwd(),"/meta_lease2.csv")) %>% column_to_rownames("sample_id")
meta_lease1 <- read_csv(paste0(getwd(),"/meta_lease1.csv")) %>% column_to_rownames("sample_id")

# common ASVs in at least 2 methods

# phytoplankton
asv_dist_phyto <- asvs_distance(pseq.euk.list.clr[[2]],ASV_common_phyto , "Phytoplankton")

phyto_asv_dist <- wrap_plots(asv_dist_phyto, nrow = 2) + plot_layout(guides = "collect") & theme(legend.position = "bottom", axis.text.x = element_text(angle = 45, size = 8))

ggsave(phyto_asv_dist, filename = paste0(folder_path, "plots/random_forest_18s/distance_asv_phyto.tiff"), compression = "lzw", width = 10, height = 10)

# Other Protists
asv_dist_other <- asvs_distance(pseq.euk.list.clr[[3]],ASV_common_other , "Other Prostists")

other_asv_dist <- wrap_plots(asv_dist_other, nrow = 2) + plot_layout(guides = "collect") & theme(legend.position = "bottom", axis.text.x = element_text(angle = 45, size = 8))

ggsave(other_asv_dist, filename = paste0(folder_path, "plots/random_forest_18s/distance_asv_other.tiff"), compression = "lzw",  width = 12, height = 10)

```


## Heatmap
```{r}
library(tidyHeatmap)
library(ComplexHeatmap)

rm(top_vip, df, ht_df, ht)

ht_vip <- function(pseq, df){

top_vip <-  df

df <- pseq %>% 
    psmelt.dplyr() %>% 
    rename(ASV = OTU)
ht_df <- df %>% filter(ASV %in% top_vip) %>% 
    mutate(ASV = str_remove(ASV, "e_asv_"),
           ASV = str_c(ASV, Phylum, Order, sep = "_"))
#  mutate(Depth = fct_relevel(Depth, c("DP1", "DP2", "DP3"))) %>%
  #group_by(site) %>%
ht <- ht_df %>%  heatmap(.column = Sample, 
          .row = ASV,
          .value = Abundance,
          column_km = 2,
          # row_km = 2,
          palette_value = circlize::colorRamp2(c(-4,  -1, 0, 1, 4), viridis::magma(5)),
          column_names_gp = grid::gpar(fontsize = 7.5),
          row_names_gp = grid::gpar(fontsize = 11, fontface = "italic"),
          column_names_rot = -45,
          cluster_rows=FALSE,
         # column_title_gp = grid::gpar(fontsize = 3),
         # palette_grouping = list(c("#66C2A5", "#FC8D62"))
          ) %>%
  add_tile(Family,palette = friendly_cols) %>%
  add_tile(Depth, palette = color_depth) %>%
  add_tile(site, palette = lease_control) 

return(ht)

}

# Phytoplankton
asvs_phyto <- phyto_asv %>% pull (ASV)
phyto_ht <- ht_vip(pseq.euk.list.clr[[2]] , asvs_phyto)

tiff(filename = paste0(folder_path,"plots/random_forest_18s/heatmap_asvs_phyto.tiff"), compression = "lzw", width = 15.8, height = 10, units = "in", res = 300)
phyto_ht
dev.off()

# Other Protists
asvs_other <- other_asv %>% pull(ASV)
other_ht <- ht_vip(pseq.euk.list.clr[[3]] , asvs_other )

tiff(filename = paste0(folder_path,"plots/random_forest_18s/heatmap_asvs_other.tiff"), compression = "lzw", width = 16, height = 12, units = "in", res = 300)
other_ht
dev.off()

```


